# Train데이터 불러오기

각 년도별 DataFrame을 리스트에 append합니다.

리스트 인덱스별로

0: 2017년도 데이터   

1: 2018년도 데이터 

2: 2019년도 데이터 

3: 202년도 데이터 

입니다

In [1]:
RANDOM_STATE=42

In [2]:
# from google.colab import drive
# drive.mount('/content/drive')

In [3]:
#cd /content/drive/Shareddrives/dacon_know/dacon_know/junghoon/

In [4]:
import pandas as pd
import numpy as np
from glob import glob
from tqdm import tqdm
import warnings

RANDOM_STATE = 42
np.seed = 42
DATA_PATH = "../data_0120/"

warnings.filterwarnings(action='ignore')
PATH_2017 = DATA_PATH + "train/KNOW_2017.csv"
PATH_2018 = DATA_PATH + "train/KNOW_2018.csv"
PATH_2019 = DATA_PATH + "train/KNOW_2019.csv"
PATH_2020 = DATA_PATH + "train/KNOW_2020.csv"

paths = [PATH_2017, PATH_2018, PATH_2019, PATH_2020]

know_train = [pd.read_csv(path) for path in paths]

In [5]:
fi2017 = pd.read_csv("../feature_importance/fi_2017.csv")
fi2018 = pd.read_csv("../feature_importance/fi_2018.csv")
fi2019 = pd.read_csv("../feature_importance/fi_2019.csv")
fi2020 = pd.read_csv("../feature_importance/fi_2020.csv")

In [6]:
fi2017.shape

(3812, 2)

In [7]:
fi2018.shape

(4560, 2)

In [8]:
fi2019.shape

(3693, 2)

In [9]:
fi2020.shape

(1234, 2)

In [10]:
fi2017.loc[fi2017['Value']>0.005]

Feature     Value
0       bq3  0.007412
1    bq41_3  0.007298
2       bq7  0.007209
3    bq41_1  0.006806
4      bq37  0.006602
..      ...       ...
114  aq22_1  0.005149
115     bq9  0.005142
116    bq27  0.005099
117     bq5  0.005034
118  aq40_2  0.005010

[119 rows x 2 columns]

In [11]:
list(fi2017.loc[fi2017['Value']>0.005, 'Feature'])

['bq3',
 'bq41_3',
 'bq7',
 'bq41_1',
 'bq37',
 'bq6',
 'bq38',
 'aq16_2',
 'bq22',
 'aq32_2',
 'aq16_1',
 'bq18_3',
 'aq29_2',
 'aq26_2',
 'aq19_2',
 'aq32_1',
 'aq41_2',
 'aq11_2',
 'aq19_1',
 'aq37_2',
 'aq10_2',
 'aq15_2',
 'aq14_2',
 'bq23',
 'aq17_2',
 'aq31_2',
 'aq2_2',
 'aq12_2',
 'aq27_2',
 'aq35_2',
 'aq11_1',
 'aq34_2',
 'aq7_2',
 'aq5_2',
 'aq1_2',
 'aq6_1',
 'aq36_2',
 'aq33_2',
 'aq1_1',
 'aq35_1',
 'aq30_2',
 'aq25_2',
 'aq6_2',
 'aq4_2',
 'aq8_2',
 'aq13_2',
 'aq24_2',
 'aq37_1',
 'aq29_1',
 'aq9_2',
 'aq30_1',
 'aq39_2',
 'aq3_2',
 'aq12_1',
 'aq5_1',
 'aq34_1',
 'aq27_1',
 'aq2_1',
 'bq8_3',
 'aq26_1',
 'aq41_1',
 'bq8_2',
 'aq17_1',
 'aq28_1',
 'aq31_1',
 'aq4_1',
 'aq28_2',
 'aq9_1',
 'aq38_2',
 'aq33_1',
 'aq7_1',
 'aq36_1',
 'aq3_1',
 'bq18_4',
 'aq25_1',
 'bq19',
 'aq38_1',
 'bq35',
 'bq5_1',
 'bq18_2',
 'aq14_1',
 'bq18_5',
 'bq16',
 'aq24_1',
 'aq8_1',
 'aq13_1',
 'bq12_1',
 'bq14',
 'aq18_2',
 'aq15_1',
 'bq18_1',
 'aq10_1',
 'aq39_1',
 'aq18_1',
 'bq4',
 'bq

In [12]:
use_cols_2017 = ['bq3','bq41_3','bq7','bq41_1','bq37','bq6','bq38','aq16_2','bq22','aq32_2','aq16_1','bq18_3',
                 'aq29_2','aq26_2','aq19_2','aq32_1','aq41_2','aq11_2','aq19_1','aq37_2','aq10_2','aq15_2',
                 'aq14_2','bq23','aq17_2','aq31_2','aq2_2','aq12_2','aq27_2','aq35_2','aq11_1','aq34_2','aq7_2',
                 'aq5_2','aq1_2','aq6_1','aq36_2','aq33_2','aq1_1','aq35_1','aq30_2','aq25_2','aq6_2','aq4_2',
                 'aq8_2','aq13_2','aq24_2','aq37_1','aq29_1','aq9_2','aq30_1','aq39_2','aq3_2','aq12_1','aq5_1',
                 'aq34_1','aq27_1','aq2_1','bq8_3','aq26_1','aq41_1','bq8_2','aq17_1','aq28_1','aq31_1','aq4_1',
                 'aq28_2','aq9_1','aq38_2','aq33_1','aq7_1','aq36_1','aq3_1','bq18_4','aq25_1','bq19','aq38_1',
                 'bq35','bq5_1','bq18_2','aq14_1','bq18_5','bq16','aq24_1','aq8_1','aq13_1','bq12_1','bq14',
                 'aq18_2','aq15_1','bq18_1','aq10_1','aq39_1','aq18_1','bq4','bq17','bq15_1','bq29','bq11',
                 'aq22_2','bq13','bq18_7','bq15_2','bq18_6','bq8_1','aq20_1','bq20','aq20_2','aq21_2',
                 'aq23_2','bq15_3','bq21','bq28','aq21_1','aq22_1','bq9','bq27','bq5','aq40_2']

In [13]:
know_train[0][use_cols_2017]

bq3  bq41_3  bq7  bq41_1  bq37  bq6  bq38  aq16_2  bq22  aq32_2  ...  \
0       6    2200    2    4000    52    5     2       6     5       0  ...   
1       3    2400    5       0    38    6     4       4     6       0  ...   
2       3    2400    5    4000    50    6     4       4     5       0  ...   
3       6    3500    4    7000    42    2     4       4     6       3  ...   
4       4    2500    5    4000    51    6     4       4     3       0  ...   
...   ...     ...  ...     ...   ...  ...   ...     ...   ...     ...  ...   
9481    2    1800    4    5200    50    3     4       2     1       3  ...   
9482    6    3000    5    4000    37    6     4       2     5       4  ...   
9483    1    1500    1    2700    32    6     2       7     1       0  ...   
9484    4    2500    2    6800    40    5     4       3     5       2  ...   
9485    5    3700    4    4000    48    3     3       4     1       4  ...   

      aq23_2  bq15_3  bq21  bq28  aq21_1  aq22_1  bq9  bq27  bq5  aq40_2  
0          4       3     3     3       3       4    4     4    1       2  
1          4       3     3     3       1       3    4     4    2       0  
2          4       3     2     2       4       3    4     4    2       0  
3          3       2     2     4       5       5    4     4    1       2  
4          4       2     2     3       3       3    4     4    2       0  
...      ...     ...   ...   ...     ...     ...  ...   ...  ...     ...  
9481       0       4     3     4       2       1    4     4    1       3  
9482       5       4     3     4       4       4    3     3    1       0  
9483       0       2     3     3       4       4    3     4    2       0  
9484       2       3     2     3       3       3    3     4    2       0  
9485       4       4     1     3       3       3    4     3    2       4  

[9486 rows x 119 columns]

In [14]:
fi2018.loc[fi2018['Value']>0.005]

Feature     Value
0        bq36  0.008806
1      bq41_2  0.008711
2         bq3  0.008505
3         bq7  0.008417
4        cq15  0.008413
..        ...       ...
78  sim_job_0  0.005648
79       cq14  0.005548
80       cq27  0.005238
81       cq20  0.005134
82     bq24_1  0.005050

[83 rows x 2 columns]

In [15]:
list(fi2018.loc[fi2018['Value']>0.005, 'Feature'])

['bq36',
 'bq41_2',
 'bq3',
 'bq7',
 'cq15',
 'cq49',
 'bq41_1',
 'cq46',
 'bq40',
 'cq10',
 'cq29',
 'cq44',
 'iq3',
 'bq20',
 'cq5',
 'bq6',
 'cq30',
 'cq40',
 'cq34',
 'cq32',
 'cq42',
 'iq4',
 'cq2',
 'cq18',
 'cq4',
 'cq8',
 'cq13',
 'cq11',
 'cq37',
 'cq36',
 'iq1',
 'cq38',
 'iq2',
 'cq12',
 'cq45',
 'cq41',
 'cq43',
 'iq6',
 'cq7',
 'cq3',
 'cq9',
 'cq39',
 'bq37',
 'iq5',
 'bq8_3',
 'bq28',
 'cq6',
 'bq11',
 'bq34',
 'cq33',
 'cq17',
 'bq21',
 'bq8_2',
 'bq5_1',
 'cq35',
 'bq12_1',
 'cq19',
 'cq16',
 'bq10',
 'cq1',
 'cq31',
 'bq8_1',
 'bq9',
 'bq12_5',
 'bq19',
 'bq4',
 'cq28',
 'bq12_4',
 'cq21',
 'able_job_0',
 'bq12_2',
 'bef_job_0',
 'bq12_3',
 'bq5',
 'ubda_cnt',
 'bq16',
 'bq35',
 'cq22',
 'sim_job_0',
 'cq14',
 'cq27',
 'cq20',
 'bq24_1']

In [30]:
use_cols_2018 = ['bq36',
 'bq41_2',
 'bq3',
 'bq7',
 'cq15',
 'cq49',
 'bq41_1',
 'cq46',
 'bq40',
 'cq10',
 'cq29',
 'cq44',
 'iq3',
 'bq20',
 'cq5',
 'bq6',
 'cq30',
 'cq40',
 'cq34',
 'cq32',
 'cq42',
 'iq4',
 'cq2',
 'cq18',
 'cq4',
 'cq8',
 'cq13',
 'cq11',
 'cq37',
 'cq36',
 'iq1',
 'cq38',
 'iq2',
 'cq12',
 'cq45',
 'cq41',
 'cq43',
 'iq6',
 'cq7',
 'cq3',
 'cq9',
 'cq39',
 'bq37',
 'iq5',
 'bq8_3',
 'bq28',
 'cq6',
 'bq11',
 'bq34',
 'cq33',
 'cq17',
 'bq21',
 'bq8_2',
 'bq5_1',
 'cq35',
 'bq12_1',
 'cq19',
 'cq16',
 'bq10',
 'cq1',
 'cq31',
 'bq8_1',
 'bq9',
 'bq12_5',
 'bq19',
 'bq4',
 'cq28',
 'bq12_4',
 'cq21',
 #'able_job',
 'bq12_2',
 #'bef_job',
 'bq12_3',
 'bq5',
 'ubda_cnt',
 'bq16',
 'bq35',
 'cq22',
 #'sim_job',
 'cq14',
 'cq27',
 'cq20',
 ]

In [17]:
know_train[1][use_cols_2018]

bq36  bq41_2  bq3  bq7  cq15  cq49  bq41_1  cq46  bq40  cq10  ...  \
0     27.0    3400    2    4     1     1    3800     2  42.0     4  ...   
1     28.0    1800    2    2     3     3    2300     3  40.0     3  ...   
2     59.0       0    1    1     2     2       0     4  30.0     2  ...   
3     38.0    2750    2    4     4     3    3350     3  40.0     3  ...   
4     27.0    2300    2    4     2     1    2500     3  40.0     2  ...   
...    ...     ...  ...  ...   ...   ...     ...   ...   ...   ...  ...   
9067  26.0       0    1    5     3     3       0     4  60.0     5  ...   
9068  56.0       0    1    5     2     1       0     4  52.0     5  ...   
9069  66.0       0    1    5     1     3       0     3  38.0     3  ...   
9070  45.0       0    1    5     1     5       0     5  56.0     3  ...   
9071  39.0    2800    2    5     1     1    4200     1  40.0     1  ...   

      bq12_3  bq5  ubda_cnt  bq16  bq35  cq22  sim_job  cq14  cq27  cq20  
0          4    1         0     2   1.0     1    치기 공사     4     3     4  
1          5    1         1     1   1.0     1        0     4     1     2  
2          2    2         0     1   1.0     2       사장     2     2     1  
3          4    2         0     2   1.0     2        0     4     3     1  
4          4    2         0     1   2.0     1        0     5     1     1  
...      ...  ...       ...   ...   ...   ...      ...   ...   ...   ...  
9067       3    1         1     1   1.0     2        0     5     1     1  
9068       3    2         1     1   1.0     1        0     5     1     1  
9069       3    1         1     1   2.0     4  비뇨기 전문의     5     1     1  
9070       3    1         0     1   1.0     1    피부 의사     5     1     1  
9071       4    2         0     2   1.0     1        0     5     1     3  

[9072 rows x 82 columns]

In [18]:
fi2019.loc[fi2019['Value']>0.005]

Feature     Value
0       bq3  0.008363
1      bq26  0.008066
2    bq31_2  0.008028
3       bq7  0.007824
4      bq30  0.007782
..      ...       ...
102  kq15_2  0.005255
103  bq12_3  0.005244
104  kq19_1  0.005200
105  kq33_2  0.005178
106  kq15_1  0.005104

[107 rows x 2 columns]

In [19]:
list(fi2019.loc[fi2019['Value']>0.005, 'Feature'])

['bq3',
 'bq26',
 'bq31_2',
 'bq7',
 'bq30',
 'bq31_1',
 'kq5_2',
 'sq16',
 'bq6',
 'kq5_1',
 'kq13_2',
 'bq17',
 'kq29_1',
 'sq15',
 'kq9_2',
 'kq13_1',
 'bq14_5',
 'kq29_2',
 'bq21_3',
 'kq32_2',
 'bq14_4',
 'bq21_1',
 'kq31_2',
 'bq16_1',
 'sq4',
 'bq14_2',
 'bq21_2',
 'sq14',
 'bq20',
 'kq2_2',
 'bq14_1',
 'kq14_2',
 'bq14_3',
 'bq27',
 'kq31_1',
 'kq23_2',
 'kq14_1',
 'kq23_1',
 'kq32_1',
 'kq2_1',
 'bq16_2',
 'bq5_1',
 'kq30_2',
 'bq19',
 'bq4',
 'sq6',
 'sq13',
 'sq1',
 'bq8_3',
 'kq4_1',
 'sq7',
 'kq25_2',
 'kq9_1',
 'bq15',
 'sq9',
 'kq18_2',
 'bq8_2',
 'bq13_1',
 'sq12',
 'bq8_1',
 'sq10',
 'sq5',
 'sq8',
 'sq11',
 'kq24_2',
 'sq2',
 'bq16_3',
 'bq11',
 'kq1_2',
 'bq13_2',
 'kq30_1',
 'bq13_3',
 'bq10',
 'bq16_5',
 'kq4_2',
 'kq25_1',
 'kq22_1',
 'sq3',
 'kq24_1',
 'bq12_1',
 'bq9',
 'kq18_1',
 'kq22_2',
 'kq3_2',
 'kq1_1',
 'bq16_4',
 'kq7_1',
 'bq12_5',
 'bq5',
 'kq10_2',
 'kq3_1',
 'kq10_1',
 'kq7_2',
 'bef_job_0',
 'bq25',
 'bq12_4',
 'kq11_2',
 'kq11_1',
 'kq19_2',
 'kq6

In [20]:
use_cols_2019 = ['bq3',
 'bq26',
 'bq31_2',
 'bq7',
 'bq30',
 'bq31_1',
 'kq5_2',
 'sq16',
 'bq6',
 'kq5_1',
 'kq13_2',
 'bq17',
 'kq29_1',
 'sq15',
 'kq9_2',
 'kq13_1',
 'bq14_5',
 'kq29_2',
 'bq21_3',
 'kq32_2',
 'bq14_4',
 'bq21_1',
 'kq31_2',
 'bq16_1',
 'sq4',
 'bq14_2',
 'bq21_2',
 'sq14',
 'bq20',
 'kq2_2',
 'bq14_1',
 'kq14_2',
 'bq14_3',
 'bq27',
 'kq31_1',
 'kq23_2',
 'kq14_1',
 'kq23_1',
 'kq32_1',
 'kq2_1',
 'bq16_2',
 'bq5_1',
 'kq30_2',
 'bq19',
 'bq4',
 'sq6',
 'sq13',
 'sq1',
 'bq8_3',
 'kq4_1',
 'sq7',
 'kq25_2',
 'kq9_1',
 'bq15',
 'sq9',
 'kq18_2',
 'bq8_2',
 'bq13_1',
 'sq12',
 'bq8_1',
 'sq10',
 'sq5',
 'sq8',
 'sq11',
 'kq24_2',
 'sq2',
 'bq16_3',
 'bq11',
 'kq1_2',
 'bq13_2',
 'kq30_1',
 'bq13_3',
 'bq10',
 'bq16_5',
 'kq4_2',
 'kq25_1',
 'kq22_1',
 'sq3',
 'kq24_1',
 'bq12_1',
 'bq9',
 'kq18_1',
 'kq22_2',
 'kq3_2',
 'kq1_1',
 'bq16_4',
 'kq7_1',
 'bq12_5',
 'bq5',
 'kq10_2',
 'kq3_1',
 'kq10_1',
 'kq7_2',
 #'bef_job',
 'bq25',
 'bq12_4',
 'kq11_2',
 'kq11_1',
 'kq19_2',
 'kq6_1',
 'kq6_2',
 'bq12_2',
 'kq15_2',
 'bq12_3',
 'kq19_1',
 'kq33_2',
 'kq15_1']

In [21]:
know_train[2][use_cols_2019]

bq3  bq26  bq31_2  bq7  bq30  bq31_1  kq5_2  sq16  bq6  kq5_1  ...  \
0       4    35    3000  4.0  40.0    3500      2     4  3.0      2  ...   
1       3    63       0  4.0  40.0       0      4     2  1.0      3  ...   
2       1    62       0  2.0  30.0       0      2     2  2.0      2  ...   
3       2    33    2400  5.0  40.0    3600      6     5  4.0      5  ...   
4       1    67       0  1.0  45.0       0      5     1  3.0      4  ...   
...   ...   ...     ...  ...   ...     ...    ...   ...  ...    ...  ...   
8550    4    41    3000  2.0  52.0    4500      3     4  4.0      2  ...   
8551    1    28    2400  2.0  60.0    3000      6     4  4.0      5  ...   
8552    1    50    1600  4.0  40.0    2400      4     2  5.0      4  ...   
8553    2    48    2800  2.0  48.0    4500      3     4  3.0      2  ...   
8554    4    29    3000  4.0  60.0    3500      0     5  6.0      1  ...   

      kq11_1  kq19_2  kq6_1  kq6_2  bq12_2  kq15_2  bq12_3  kq19_1  kq33_2  \
0          4       2      3      4       5       5       4       2       5   
1          1       0      5      6       3       0       3       1       4   
2          1       0      1      0       3       0       3       1       0   
3          3       4      4      5       3       4       3       3       3   
4          1       0      1      0       3       0       3       1       0   
...      ...     ...    ...    ...     ...     ...     ...     ...     ...   
8550       3       3      2      3       3       3       4       2       3   
8551       1       0      1      0       3       0       5       1       0   
8552       2       2      1      0       3       2       4       2       2   
8553       3       0      2      4       3       3       3       1       4   
8554       1       0      1      0       3       4       3       1       0   

      kq15_1  
0          4  
1          1  
2          1  
3          3  
4          1  
...      ...  
8550       2  
8551       1  
8552       2  
8553       3  
8554       3  

[8555 rows x 107 columns]

In [22]:
fi2020.loc[fi2020['Value']>0.005]

Feature     Value
0    bq30_2  0.007186
1       bq7  0.007130
2       bq3  0.006464
3      bq29  0.006262
4       bq6  0.006096
..      ...       ...
94      vq3  0.005044
95  saq27_2  0.005037
96  saq14_1  0.005032
97   bq21_1  0.005023
98  saq32_1  0.005006

[99 rows x 2 columns]

In [23]:
list(fi2020.loc[fi2020['Value']>0.005, 'Feature'])

['bq30_2',
 'bq7',
 'bq3',
 'bq29',
 'bq6',
 'bq26',
 'saq3_2',
 'saq41_2',
 'bq30_1',
 'saq44_2',
 'saq31_2',
 'saq43_2',
 'bq4',
 'saq38_1',
 'saq7_2',
 'saq38_2',
 'saq40_2',
 'saq31_1',
 'saq42_2',
 'bq25',
 'bq22_1',
 'saq3_1',
 'saq5_2',
 'saq1_2',
 'saq42_1',
 'bq23_1',
 'saq20_2',
 'saq41_1',
 'saq40_1',
 'saq7_1',
 'saq36_2',
 'saq14_2',
 'bq22_4',
 'bq22_2',
 'saq43_1',
 'saq10_2',
 'saq10_1',
 'saq24_2',
 'saq4_2',
 'saq18_2',
 'saq44_1',
 'saq4_1',
 'bq22_3',
 'saq13_2',
 'saq25_2',
 'saq19_1',
 'saq23_2',
 'saq5_1',
 'saq20_1',
 'bq20',
 'saq34_2',
 'bq23_2',
 'saq22_2',
 'saq11_2',
 'saq39_2',
 'bq22_5',
 'vq6',
 'saq19_2',
 'saq15_2',
 'saq9_2',
 'saq1_1',
 'saq26_1',
 'saq35_2',
 'saq23_1',
 'saq2_2',
 'saq16_2',
 'saq32_2',
 'saq24_1',
 'bq21_2',
 'bq23_3',
 'saq6_2',
 'saq37_2',
 'saq11_1',
 'saq8_2',
 'saq36_1',
 'saq17_2',
 'saq33_2',
 'saq12_2',
 'saq16_1',
 'saq26_2',
 'saq29_1',
 'bq8_3',
 'saq18_1',
 'saq22_1',
 'saq27_1',
 'vq10',
 'bq22_6',
 'saq6_1',
 'bq14_5

In [24]:
use_cols_2020 = ['bq30_2',
 'bq7',
 'bq3',
 'bq29',
 'bq6',
 'bq26',
 'saq3_2',
 'saq41_2',
 'bq30_1',
 'saq44_2',
 'saq31_2',
 'saq43_2',
 'bq4',
 'saq38_1',
 'saq7_2',
 'saq38_2',
 'saq40_2',
 'saq31_1',
 'saq42_2',
 'bq25',
 'bq22_1',
 'saq3_1',
 'saq5_2',
 'saq1_2',
 'saq42_1',
 'bq23_1',
 'saq20_2',
 'saq41_1',
 'saq40_1',
 'saq7_1',
 'saq36_2',
 'saq14_2',
 'bq22_4',
 'bq22_2',
 'saq43_1',
 'saq10_2',
 'saq10_1',
 'saq24_2',
 'saq4_2',
 'saq18_2',
 'saq44_1',
 'saq4_1',
 'bq22_3',
 'saq13_2',
 'saq25_2',
 'saq19_1',
 'saq23_2',
 'saq5_1',
 'saq20_1',
 'bq20',
 'saq34_2',
 'bq23_2',
 'saq22_2',
 'saq11_2',
 'saq39_2',
 'bq22_5',
 'vq6',
 'saq19_2',
 'saq15_2',
 'saq9_2',
 'saq1_1',
 'saq26_1',
 'saq35_2',
 'saq23_1',
 'saq2_2',
 'saq16_2',
 'saq32_2',
 'saq24_1',
 'bq21_2',
 'bq23_3',
 'saq6_2',
 'saq37_2',
 'saq11_1',
 'saq8_2',
 'saq36_1',
 'saq17_2',
 'saq33_2',
 'saq12_2',
 'saq16_1',
 'saq26_2',
 'saq29_1',
 'bq8_3',
 'saq18_1',
 'saq22_1',
 'saq27_1',
 'vq10',
 'bq22_6',
 'saq6_1',
 'bq14_5',
 'saq39_1',
 'saq28_2',
 'saq21_2',
 'saq35_1',
 'saq33_1',
 'vq3',
 'saq27_2',
 'saq14_1',
 'bq21_1',
 'saq32_1']

In [25]:
know_train[3][use_cols_2020]

bq30_2  bq7  bq3  bq29  bq6  bq26  saq3_2  saq41_2  bq30_1  saq44_2  \
0       2600    5    3    40    3     4       5        3    4000        4   
1       3000    5    2    40    4     4       5        5    4000        4   
2       3000    5    4    48    5     4       4        4    4300        4   
3       3300    5    3    40    4     4       3        3    3500        1   
4       2700    5    2    40    3     4       5        5    5700        2   
...      ...  ...  ...   ...  ...   ...     ...      ...     ...      ...   
8117    4000    6    5    40    3     6       3        3    5100        0   
8118    2600    5    2    45    2     4       4        0    3100        4   
8119    3000    5    3    40    3     4       4        0    8000        2   
8120    3200    5    2    40    6     5       5        0    5700        4   
8121    2700    4    2    46    4     3       5        6    3700        4   

      ...  saq39_1  saq28_2  saq21_2  saq35_1  saq33_1  vq3  saq27_2  saq14_1  \
0     ...        3        0        3        3        1    5        0        3   
1     ...        3        3        5        3        2    3        3        4   
2     ...        4        5        4        4        4    4        5        4   
3     ...        2        0        5        1        1    4        4        3   
4     ...        4        6        5        4        3    5        5        4   
...   ...      ...      ...      ...      ...      ...  ...      ...      ...   
8117  ...        3        4        4        3        2    2        4        2   
8118  ...        1        0        3        1        1    3        0        4   
8119  ...        1        0        5        1        2    3        2        2   
8120  ...        1        4        5        3        3    3        4        3   
8121  ...        4        4        5        3        3    2        4        3   

      bq21_1  saq32_1  
0          3        4  
1          3        3  
2          3        3  
3          3        1  
4          2        3  
...      ...      ...  
8117       3        1  
8118       2        1  
8119       2        1  
8120       3        3  
8121       2        4  

[8122 rows x 99 columns]

## Train Test Split

In [26]:
from sklearn.model_selection import train_test_split

In [27]:
X_train_2017, X_valid_2017, y_train_2017, y_valid_2017 = train_test_split(know_train[0][use_cols_2017], know_train[0]['knowcode'],
                                                                        test_size=0.2, random_state=RANDOM_STATE)

In [28]:
from sklearn.ensemble import ExtraTreesClassifier

rf2017 = ExtraTreesClassifier(n_estimators=100, random_state = RANDOM_STATE)
rf2017.fit(X_train_2017, y_train_2017)

ExtraTreesClassifier(random_state=42)

In [29]:
pred_2017 = rf2017.predict(X_valid_2017)

In [30]:
from sklearn.metrics import f1_score
f1_score(y_valid_2017, pred_2017, average='macro')

0.40459698199877714

## Testset 불러오기

마찬가지로 년도별로 DataFrame으로 불러온 후 리스트에 할당합니다.


In [26]:
TEST_PATH_2017 = DATA_PATH + "test/KNOW_2017_test.csv"
TEST_PATH_2018 = DATA_PATH + "test/KNOW_2018_test.csv"
TEST_PATH_2019 = DATA_PATH + "test/KNOW_2019_test.csv"
TEST_PATH_2020 = DATA_PATH + "test/KNOW_2020_test.csv"

TEST_PATHs = [TEST_PATH_2017, TEST_PATH_2018, TEST_PATH_2019, TEST_PATH_2020]

In [27]:
know_test = [pd.read_csv(path) for path in TEST_PATHs]
know_test[0].head() # 2017년도 test 샘플

idx  aq1_1  aq1_2  aq2_1  aq2_2  aq3_1  aq3_2  aq4_1  aq4_2  aq5_1  ...  \
0    0      3      4      2      2      3      3      1      0      3  ...   
1    1      5      5      3      5      5      5      5      5      4  ...   
2    2      5      5      5      4      5      4      1      0      1  ...   
3    3      4      5      5      6      4      6      3      4      4  ...   
4    4      5      6      4      5      4      5      1      0      1  ...   

   기술  컴퓨터  청진기  자동화  생산  안전  계측기  해외  작업  ubda_cnt  
0   0    1    0    0   0   0    0   0   0         1  
1   0    0    0    0   0   0    0   0   0         0  
2   0    0    0    0   0   0    0   0   0         0  
3   0    1    0    0   0   0    0   0   0         1  
4   0    0    0    0   0   0    0   0   0         1  

[5 rows x 304 columns]

In [28]:
from sklearn.ensemble import RandomForestClassifier

In [29]:
rf2017 = RandomForestClassifier(n_estimators=300, random_state = RANDOM_STATE)
rf2017.fit(know_train[0][use_cols_2017], know_train[0]['knowcode'])

RandomForestClassifier(n_estimators=300, random_state=42)

In [30]:
pred_2017 = rf2017.predict(know_test[0][use_cols_2017])

In [32]:
import pickle
with open('rf_300_pred2017.pkl', 'wb') as f:
    pickle.dump(list(pred_2017), f)

In [31]:
rf2018 = RandomForestClassifier(n_estimators=300, random_state = RANDOM_STATE)
rf2018.fit(know_train[1][use_cols_2018], know_train[1]['knowcode'])

RandomForestClassifier(n_estimators=300, random_state=42)

In [32]:
pred_2018 = rf2018.predict(know_test[1][use_cols_2018])

In [33]:
import pickle
with open('rf_300_pred2018.pkl', 'wb') as f:
    pickle.dump(list(pred_2018), f)

In [ ]:
rf2019 = RandomForestClassifier(n_estimators=300, random_state = RANDOM_STATE)
rf2019.fit(know_train[2][use_cols_2019], know_train[2]['knowcode'])

In [ ]:
pred_2019 = rf2019.predict(know_test[2][use_cols_2019])

In [ ]:
import pickle
with open('rf_300_pred2019.pkl', 'wb') as f:
    pickle.dump(list(pred_2019), f)

In [ ]:
rf2020 = RandomForestClassifier(n_estimators=300, random_state = RANDOM_STATE)
rf2020.fit(know_train[3][use_cols_2020], know_train[3]['knowcode'])

In [ ]:
pred_2020 = rf2020.predict(know_test[3][use_cols_2020])

In [ ]:
import pickle
with open('rf_300_pred2020.pkl', 'wb') as f:
    pickle.dump(list(pred_2020), f)

In [34]:
with open('rf_300_pred2017.pkl', 'rb') as f:
    pred2017 = pickle.load(f)

In [35]:
pred2017_df = pd.DataFrame(pred2017)

In [38]:
pred2017_df.to_csv('name2017.csv',index=False)

In [39]:
with open('rf_300_pred2018.pkl', 'rb') as f:
    pred2018 = pickle.load(f)

In [40]:
pred2018_df = pd.DataFrame(pred2018)

In [41]:
pred2018_df.to_csv('name2018.csv',index=False)

In [ ]:
with open('rf_300_pred2019.pkl', 'rb') as f:
    pred2019 = pickle.load(f)

In [ ]:
with open('rf_300_pred2020.pkl', 'rb') as f:
    pred2020 = pickle.load(f)

## Option 0: 모든 문항 사용하되, ET의 feature importance 순으로 자르기

In [2]:
from pycaret.classification import *

### 2017

In [11]:
numeric_cols_2017 = ['aq1_1', 'aq1_2', 'aq2_1', 'aq2_2', 'aq3_1', 'aq3_2', 'aq4_1', 'aq4_2', 'aq5_1', 'aq5_2', 
                     'aq6_1', 'aq6_2', 'aq7_1', 'aq7_2', 'aq8_1', 'aq8_2', 'aq9_1', 'aq9_2', 'aq10_1', 'aq10_2',
                    'aq11_1', 'aq11_2', 'aq12_1', 'aq12_2', 'aq13_1', 'aq13_2', 'aq14_1', 'aq14_2', 'aq15_1', 'aq15_2', 
                     'aq16_1', 'aq16_2', 'aq17_1', 'aq17_2', 'aq18_1', 'aq18_2', 'aq19_1', 'aq19_2', 'aq20_1', 'aq20_2',
                    'aq21_1', 'aq21_2', 'aq22_1', 'aq22_2', 'aq23_1', 'aq23_2', 'aq24_1', 'aq24_2', 'aq25_1', 'aq25_2', 
                     'aq26_1', 'aq26_2', 'aq27_1', 'aq27_2', 'aq28_1', 'aq28_2', 'aq29_1', 'aq29_2', 'aq30_1', 'aq30_2',
                    'aq31_1', 'aq31_2', 'aq32_1', 'aq32_2', 'aq33_1', 'aq33_2', 'aq34_1', 'aq34_2', 'aq35_1', 'aq35_2', 
                     'aq36_1', 'aq36_2', 'aq37_1', 'aq37_2', 'aq38_1', 'aq38_2', 'aq39_1', 'aq39_2', 'aq40_1', 'aq40_2',
                     'aq41_1', 'aq41_2', 'bq3', 'bq5_1', 'bq7', 'bq8_1', 'bq8_2', 'bq8_3', 'bq9', 'bq10',
                     'bq11', 'bq12_1', 'bq12_5', 'bq13', 'bq14', 'bq15_1', 'bq15_2', 'bq15_3', 'bq16', 'bq17', 
                     'bq18_1', 'bq18_2', 'bq18_3', 'bq18_4', 'bq18_5', 'bq18_6', 'bq18_7', 'bq19', 'bq20', 'bq21', 
                     'bq22', 'bq25', 'bq26', 'bq27', 'bq28', 'bq29', 'bq35', 'bq37', 'bq38', ]
numeric_pure_cols_2017 = ['bq23', 'bq37', 'bq41_1', 'bq41_2', 'bq41_3', 'ubda_cnt']
categorical_cols_2017 = ['bq1', 'bq2', 'bq39_1', 'bq39_2', 'bq6_mark','bq12_2_mark','bq12_3_mark','bq12_4_mark',]
binary_cols_2017 = ['bq4', 'bq5', 'bq24_1', 'bq24_2', 'bq24_3', 'bq24_4', 'bq24_5', 'bq24_6', 'bq24_7', 'bq24_8', 'bq36', 'bq40', ]
half_cols_2017 = ['bq6', 'bq12_2', 'bq12_3', 'bq12_4', ]
text_cols_2017 = ['bq4_1a', 'bq4_1b', 'bq4_1c', 'bq5_2', 'bq19_1', 'bq30', 'bq31', 'bq32', 'bq33', 'bq34', 'bq38_1',]

In [12]:
all_cols_2017 = numeric_cols_2017 + numeric_pure_cols_2017\
                + categorical_cols_2017 + binary_cols_2017 + half_cols_2017 + text_cols_2017

In [13]:
all_nontext_cols_2017 = numeric_cols_2017 + numeric_pure_cols_2017\
                + categorical_cols_2017 + binary_cols_2017 + half_cols_2017

In [14]:
no_match_cols_2017 = [col for col in know_train[0].columns if col not in all_cols_2017]
no_match_cols_2017

['idx',
 'sim_job',
 'bef_job',
 'able_job',
 'major',
 'description',
 '제품',
 '기기',
 '인구',
 '줄자',
 '공급',
 '정보처리기사',
 '필기',
 '오토캐드',
 '포화상태',
 '설계',
 '음악',
 '회계',
 '카메라',
 '서비스',
 '디자인',
 '전문의',
 '사업',
 '무대',
 '고객',
 '직종',
 '훈련',
 '축소',
 '지식',
 '안전교육',
 '파악',
 '변화',
 '드릴',
 '침체',
 '자동',
 '기계',
 '현미경',
 '건축',
 '장갑',
 '에너지',
 '정보',
 '캐드',
 '기록',
 '정부',
 '무전기',
 '편집',
 '가위',
 '기구',
 '마이크',
 '의상',
 '치료',
 '프로그램',
 '인공',
 '업체',
 '전기',
 '교육',
 '향상',
 '보안',
 '분석',
 '수가',
 '드라이버',
 '직무교육',
 '문화',
 '자료',
 '실무교육',
 '변동',
 '인식',
 '시장',
 '로봇',
 '검사',
 '전화기',
 '노인',
 '엑셀',
 '건설',
 '병원',
 '면허증',
 '향후',
 '기피',
 '노트북',
 '포토샵',
 '단기',
 '보험',
 '지능',
 '판매',
 '피아노',
 '자격증',
 '볼펜',
 '시스템',
 '기업',
 '운전면허증',
 '회사',
 '인원',
 '환자',
 '전산',
 '오피스',
 '화로',
 '망치',
 '복지',
 '감축',
 '간호사',
 '활동',
 '매체',
 '시험',
 '연구',
 '교원자격증',
 '한글',
 '시설',
 '생활',
 '교재',
 '영향',
 '특성',
 '측정',
 '정책',
 '경기',
 '의료',
 '공구',
 '활성화',
 '고령화',
 '배출',
 '설비',
 '방법',
 '인터넷',
 '관리',
 '실험',
 '측정기',
 '불황',
 '방송',
 '지속',
 '의사면허증',
 '전기기사',
 '건강',
 '제조

In [15]:
korean_cols_2017 = [col for col in no_match_cols_2017 if not col in ['idx','sim_job', 'bef_job',
                                                                     'able_job','major','description','knowcode']] 

In [16]:
korean_cols_2017

['제품',
 '기기',
 '인구',
 '줄자',
 '공급',
 '정보처리기사',
 '필기',
 '오토캐드',
 '포화상태',
 '설계',
 '음악',
 '회계',
 '카메라',
 '서비스',
 '디자인',
 '전문의',
 '사업',
 '무대',
 '고객',
 '직종',
 '훈련',
 '축소',
 '지식',
 '안전교육',
 '파악',
 '변화',
 '드릴',
 '침체',
 '자동',
 '기계',
 '현미경',
 '건축',
 '장갑',
 '에너지',
 '정보',
 '캐드',
 '기록',
 '정부',
 '무전기',
 '편집',
 '가위',
 '기구',
 '마이크',
 '의상',
 '치료',
 '프로그램',
 '인공',
 '업체',
 '전기',
 '교육',
 '향상',
 '보안',
 '분석',
 '수가',
 '드라이버',
 '직무교육',
 '문화',
 '자료',
 '실무교육',
 '변동',
 '인식',
 '시장',
 '로봇',
 '검사',
 '전화기',
 '노인',
 '엑셀',
 '건설',
 '병원',
 '면허증',
 '향후',
 '기피',
 '노트북',
 '포토샵',
 '단기',
 '보험',
 '지능',
 '판매',
 '피아노',
 '자격증',
 '볼펜',
 '시스템',
 '기업',
 '운전면허증',
 '회사',
 '인원',
 '환자',
 '전산',
 '오피스',
 '화로',
 '망치',
 '복지',
 '감축',
 '간호사',
 '활동',
 '매체',
 '시험',
 '연구',
 '교원자격증',
 '한글',
 '시설',
 '생활',
 '교재',
 '영향',
 '특성',
 '측정',
 '정책',
 '경기',
 '의료',
 '공구',
 '활성화',
 '고령화',
 '배출',
 '설비',
 '방법',
 '인터넷',
 '관리',
 '실험',
 '측정기',
 '불황',
 '방송',
 '지속',
 '의사면허증',
 '전기기사',
 '건강',
 '제조업',
 '학생',
 '워드',
 '개발',
 '대본',
 '접기',
 '통계',
 '자동차',
 '기계화',
 '일러스트',


In [17]:
know_train[0]['knowcode']

0       825101
1       140204
2       140204
3       140601
4       140204
         ...  
9481    411301
9482    151105
9483    701101
9484     25402
9485     15201
Name: knowcode, Length: 9486, dtype: int64

In [18]:
clf2017 = setup(know_train[0], target = 'knowcode',
            fold=5,
            numeric_features = numeric_cols_2017 + numeric_pure_cols_2017 + binary_cols_2017 + half_cols_2017,
            categorical_features = categorical_cols_2017 + korean_cols_2017,
            ignore_features = ['idx', 'description'],
            session_id=RANDOM_STATE,
            silent=True,
            #feature_selection=True,
            #feature_selection_threshold=0.5,
           )

In [19]:
et_2017 = create_model('et')

In [27]:
#evaluate_model(et_2017)

In [22]:
fi_2017 = pd.DataFrame({'Feature': get_config('X_train').columns, 'Value' : abs(et_2017.feature_importances_)}).sort_values(by='Value', ascending=False)

In [26]:
fi_2017

Feature     Value
82                     bq3  0.007412
138                 bq41_3  0.007298
87                     bq7  0.007209
136                 bq41_1  0.006806
133                   bq37  0.006602
...                    ...       ...
3193            major_영상연출  0.000000
1755  bef_job_학원강사구매 매입관리직  0.000000
3636           major_화학교육학  0.000000
1779          bef_job_호텔리어  0.000000
2708       major_교육학(상담심리)  0.000000

[3812 rows x 2 columns]

In [54]:
fi_2017.to_csv('fi_2017.csv',encoding='utf-8', index=False)

### 2018

In [5]:
numeric_cols_2018 = ['cq1', 'cq2', 'cq3', 'cq4', 'cq5', 'cq6', 'cq7', 'cq8', 'cq9', 'cq10', 
                    'cq11', 'cq12', 'cq13', 'cq14', 'cq15', 'cq16', 'cq17', 'cq18', 'cq19', 'cq20', 
                    'cq21', 'cq22', 'cq23', 'cq24', 'cq25', 'cq26', 'cq27', 'cq28', 'cq29', 'cq30',
                    'cq31', 'cq32', 'cq33', 'cq34', 'cq35', 'cq36', 'cq37', 'cq38', 'cq39', 'cq40', 
                    'cq41', 'cq42', 'cq43', 'cq44', 'cq45', 'cq46', 'cq47', 'cq49', 'cq50_1', 'cq50_2', 'cq50_3', 'cq50_4', 'cq50_5', 'cq50_6', 'cq50_7', 'cq50_8',
                    'iq1', 'iq2', 'iq3', 'iq4', 'iq5', 'iq6',
                    'bq3', 'bq5_1', 'bq7', 'bq8_1', 'bq8_2', 'bq8_3', 'bq9', 'bq10', 'bq11', 'bq12_1', 'bq12_5', 'bq18', 'bq19', 'bq20', 'bq25', 'bq26_1a', 
                     'bq26_2a', 'bq26_3a', 'bq26_4a', 'bq28', 'bq34', 'bq37', ]
numeric_pure_cols_2018 = ['bq21', 'bq36', 'bq40', 'bq41_1', 'bq41_2', 'bq41_3', 'ubda_cnt',]
categorical_cols_2018 = ['bq1', 'bq2', 'bq13', 'bq15', 'bq17', 'bq22', 'bq23', 'bq24', 'bq26_1', 'bq26_2',
                         'bq26_3', 'bq26_4', 'bq38', 'bq38_1', 'bq38_2', 'bq6_mark','bq12_2_mark','bq12_3_mark',
                         'bq12_4_mark',]
binary_cols_2018 = ['cq48', 'bq4', 'bq5', 'bq14', 'bq16', 'bq25_1', 'bq27', 'bq35', 'bq39', ]
half_cols_2018 = ['bq6', 'bq12_2', 'bq12_3', 'bq12_4', ]
text_cols_2018 = ['bq4_1a', 'bq4_1b', 'bq4_1c', 'bq5_2', 'bq28_1', 'bq29', 'bq30', 'bq31', 'bq32', 'bq33', 'bq37_1', ]

In [6]:
all_cols_2018 = numeric_cols_2018 + numeric_pure_cols_2018\
                + categorical_cols_2018 + binary_cols_2018 + half_cols_2018 + text_cols_2018

In [7]:
all_nontext_cols_2018 = numeric_cols_2018 + numeric_pure_cols_2018\
                + categorical_cols_2018 + binary_cols_2018 + half_cols_2018

In [8]:
no_match_cols_2018 = [col for col in know_train[1].columns if col not in all_cols_2018]
no_match_cols_2018

['idx',
 'sim_job',
 'bef_job',
 'able_job',
 'major',
 'description',
 '제품',
 '기기',
 '인구',
 '줄자',
 '공급',
 '정보처리기사',
 '오토캐드',
 '여가',
 '포화상태',
 '스패너',
 '설계',
 '프로그래밍',
 '카메라',
 '서비스',
 '디자인',
 '전문의',
 '사업',
 '고객',
 '직종',
 '훈련',
 '수첩',
 '축소',
 '작성',
 'MS오피스',
 '지식',
 '안전교육',
 '변화',
 '건물',
 '드릴',
 '침체',
 '기계',
 '현미경',
 '건축',
 'ERP',
 '장갑',
 '에너지',
 '각종',
 '캐드',
 '무전기',
 '정부',
 '가위',
 '마이크',
 '프로그램',
 '인공',
 '업체',
 '2급',
 '전기',
 '교육',
 '향상',
 '분석',
 '수가',
 '드라이버',
 '직무교육',
 '문화',
 '실무교육',
 '필기도구',
 '변동',
 '시장',
 '자체',
 '처리',
 '도마',
 '노인',
 '엑셀',
 '건설',
 '산업기사',
 '면허증',
 '빅데이터',
 '기피',
 '노트북',
 '포토샵',
 '보험',
 '더존',
 '지능',
 '판매',
 '사회복지사',
 '자격증',
 '볼펜',
 '시스템',
 '기업',
 '운전면허증',
 '회사',
 '인원',
 '전산',
 '오피스',
 '제작',
 '화로',
 '망치',
 '복지',
 '감축',
 '가능성',
 '온라인',
 '연구',
 '등등',
 '한글',
 '시설',
 '생활',
 '정책',
 '경기',
 'PC',
 '의료',
 '공구',
 '활성화',
 'MS',
 '고령화',
 '현장',
 '건축기사',
 '절단기',
 '설비',
 '방법',
 '해당없음',
 '인터넷',
 '관리',
 '사내',
 '불황',
 '방송',
 '기사',
 '지속',
 '의사면허증',
 '전기기사',
 'CAD',
 '토목기사',
 '요구',
 '건강'

In [9]:
korean_cols_2018 = [col for col in no_match_cols_2018 if not col in ['idx','sim_job', 'bef_job',
                                                                     'able_job','major','description','knowcode']] 

In [10]:
korean_cols_2018

['제품',
 '기기',
 '인구',
 '줄자',
 '공급',
 '정보처리기사',
 '오토캐드',
 '여가',
 '포화상태',
 '스패너',
 '설계',
 '프로그래밍',
 '카메라',
 '서비스',
 '디자인',
 '전문의',
 '사업',
 '고객',
 '직종',
 '훈련',
 '수첩',
 '축소',
 '작성',
 'MS오피스',
 '지식',
 '안전교육',
 '변화',
 '건물',
 '드릴',
 '침체',
 '기계',
 '현미경',
 '건축',
 'ERP',
 '장갑',
 '에너지',
 '각종',
 '캐드',
 '무전기',
 '정부',
 '가위',
 '마이크',
 '프로그램',
 '인공',
 '업체',
 '2급',
 '전기',
 '교육',
 '향상',
 '분석',
 '수가',
 '드라이버',
 '직무교육',
 '문화',
 '실무교육',
 '필기도구',
 '변동',
 '시장',
 '자체',
 '처리',
 '도마',
 '노인',
 '엑셀',
 '건설',
 '산업기사',
 '면허증',
 '빅데이터',
 '기피',
 '노트북',
 '포토샵',
 '보험',
 '더존',
 '지능',
 '판매',
 '사회복지사',
 '자격증',
 '볼펜',
 '시스템',
 '기업',
 '운전면허증',
 '회사',
 '인원',
 '전산',
 '오피스',
 '제작',
 '화로',
 '망치',
 '복지',
 '감축',
 '가능성',
 '온라인',
 '연구',
 '등등',
 '한글',
 '시설',
 '생활',
 '정책',
 '경기',
 'PC',
 '의료',
 '공구',
 '활성화',
 'MS',
 '고령화',
 '현장',
 '건축기사',
 '절단기',
 '설비',
 '방법',
 '해당없음',
 '인터넷',
 '관리',
 '사내',
 '불황',
 '방송',
 '기사',
 '지속',
 '의사면허증',
 '전기기사',
 'CAD',
 '토목기사',
 '요구',
 '건강',
 '재봉틀',
 '학생',
 '워드',
 '개발',
 '대본',
 '자동차',
 '기계화',
 '일러스트',
 '현상',
 

In [11]:
clf2018 = setup(know_train[1], target = 'knowcode',
            fold=5,
            numeric_features = numeric_cols_2018 + numeric_pure_cols_2018 + binary_cols_2018 + half_cols_2018,
            categorical_features = categorical_cols_2018 + korean_cols_2018,
            ignore_features = ['idx', 'description'],
            session_id=RANDOM_STATE,
            silent=True,
            #feature_selection=True,
            #feature_selection_threshold=0.5,
           )

In [12]:
et_2018 = create_model('et')

In [27]:
#evaluate_model(et_2017)

In [13]:
fi_2018 = pd.DataFrame({'Feature': get_config('X_train').columns, 'Value' : abs(et_2018.feature_importances_)}).sort_values(by='Value', ascending=False)

In [14]:
fi_2018

Feature     Value
96                   bq36  0.008806
100                bq41_2  0.008711
63                    bq3  0.008505
68                    bq7  0.008417
14                   cq15  0.008413
...                   ...       ...
4409        bq12_2_mark_0  0.000000
3932          major_유통마케팅  0.000000
1611   bef_job_자영업-의류 판매원  0.000000
1596      bef_job_자동차영업사원  0.000000
2091  able_job_공공기관 기술공무원  0.000000

[4560 rows x 2 columns]

In [15]:
fi_2018.to_csv('fi_2018.csv',encoding='utf-8', index=False)

### 2019

In [19]:
numeric_cols_2019 = ['sq1', 'sq2', 'sq3', 'sq4', 'sq5', 'sq6', 'sq7', 'sq8', 'sq9', 'sq10', 
                     'sq11', 'sq12', 'sq13', 'sq14', 'sq15', 'sq16',
                    'kq1_1', 'kq1_2', 'kq2_1', 'kq2_2', 'kq3_1', 'kq3_2', 'kq4_1', 'kq4_2', 'kq5_1', 'kq5_2', 
                     'kq6_1', 'kq6_2', 'kq7_1', 'kq7_2', 'kq8_1', 'kq8_2', 'kq9_1', 'kq9_2', 'kq10_1', 'kq10_2',
                    'kq11_1', 'kq11_2', 'kq12_1', 'kq12_2', 'kq13_1', 'kq13_2', 'kq14_1', 'kq14_2', 'kq15_1', 'kq15_2',
                     'kq16_1', 'kq16_2', 'kq17_1', 'kq17_2', 'kq18_1', 'kq18_2', 'kq19_1', 'kq19_2', 'kq20_1', 'kq20_2',
                    'kq21_1', 'kq21_2', 'kq22_1', 'kq22_2', 'kq23_1', 'kq23_2', 'kq24_1', 'kq24_2', 'kq25_1', 'kq25_2', 
                     'kq26_1', 'kq26_2', 'kq27_1', 'kq27_2', 'kq28_1', 'kq28_2', 'kq29_1', 'kq29_2', 'kq30_1', 'kq30_2',
                    'kq31_1', 'kq31_2', 'kq32_1', 'kq32_2', 'kq33_1', 'kq33_2', 
                    'bq3', 'bq5_1', 'bq7', 'bq8_1', 'bq8_2', 'bq8_3', 'bq9', 'bq10', 'bq11', 'bq12_1', 'bq12_5', 
                     'bq13_1', 'bq13_2', 'bq13_3', 'bq14_1', 'bq14_2', 'bq14_3', 'bq14_4', 'bq14_5', 'bq15', 
                     'bq16_1', 'bq16_2', 'bq16_3', 'bq16_4', 'bq16_5', 'bq17', 'bq19', 'bq20', 'bq21_1', 'bq21_2', 'bq21_3', 'bq27',]
numeric_pure_cols_2019 = ['bq26', 'bq30', 'bq31_1', 'bq31_2', 'bq31_3', 'ubda_cnt']
categorical_cols_2019 = ['bq1', 'bq2', 'bq28', 'bq28_1', 'bq28_2', 'bq6_mark','bq12_2_mark','bq12_3_mark','bq12_4_mark',]
binary_cols_2019 = ['bq4', 'bq5', 'bq18_1', 'bq18_2', 'bq18_3', 'bq18_4', 'bq18_5', 'bq18_6', 'bq18_7', 
                    'bq18_8', 'bq18_9', 'bq25', 'bq29', ]
half_cols_2019 = ['bq6', 'bq12_2', 'bq12_3', 'bq12_4', ]
text_cols_2019 = ['bq4_1a', 'bq4_1b', 'bq4_1c', 'bq5_2', 'bq18_10', 'bq20_1', 'bq22', 'bq23', 'bq24', 'bq27_1',]

In [20]:
all_cols_2019 = numeric_cols_2019 + numeric_pure_cols_2019\
                + categorical_cols_2019 + binary_cols_2019 + half_cols_2019 + text_cols_2019

In [21]:
all_nontext_cols_2019 = numeric_cols_2019 + numeric_pure_cols_2019\
                + categorical_cols_2019 + binary_cols_2019 + half_cols_2019

In [22]:
no_match_cols_2019 = [col for col in know_train[2].columns if col not in all_cols_2019]
no_match_cols_2019

['idx',
 'bef_job',
 'able_job',
 'major',
 'description',
 '제품',
 '기기',
 '출시',
 '인구',
 '정보처리기사',
 '소프트웨어',
 '설계',
 '프로그래밍',
 '적용',
 '서비스',
 '디자인',
 '필요성',
 '전문의',
 '데이터',
 '사업',
 '고객',
 '직종',
 '훈련',
 '축소',
 '작성',
 '지식',
 '안전교육',
 '파악',
 '변화',
 '개선',
 '침체',
 '정비',
 '자동',
 '항공기',
 '기계',
 '건축',
 '판단',
 '정보',
 '교체',
 '디지털',
 '직접',
 '편집',
 '기계로',
 '영상',
 '치료',
 '프로그램',
 '인공',
 '업체',
 '전기',
 '교육',
 '향상',
 '분석',
 '구조조정',
 '수가',
 '직무교육',
 '개인',
 '문화',
 '스트레스',
 '미디어',
 '실무교육',
 '처리',
 '시장',
 '로봇',
 '신규',
 '검사',
 '위한',
 '효율',
 '건설',
 '조사',
 '병원',
 '산업기사',
 '수준',
 '면허증',
 '빅데이터',
 '상담',
 '영역',
 '지능',
 '관리자',
 '판매',
 '재교육',
 '자격증',
 '의사',
 '사회복지사',
 '시스템',
 '구조',
 '기업',
 '운전면허증',
 '기본',
 '회사',
 '인원',
 '환자',
 '전산',
 '제작',
 '화로',
 '감축',
 '매체',
 '온라인',
 '연구',
 '드론',
 '시설',
 '영향',
 '경기',
 '의료',
 '활성화',
 '새로운',
 '고령화',
 '현장',
 '설비',
 '방법',
 '인터넷',
 '관리',
 '통신',
 '불황',
 '방송',
 '기사',
 '지속',
 '의사면허증',
 '전기기사',
 '가능',
 '요구',
 '토목기사',
 '건강',
 '제조업',
 '학생',
 '개발',
 '자동차',
 '기계화',
 '경제',
 '마케팅',
 '전문',
 '예술

In [23]:
korean_cols_2019 = [col for col in no_match_cols_2019 if not col in ['idx','bef_job',
                                                                     'able_job','major','description','knowcode']] 

In [24]:
korean_cols_2019

['제품',
 '기기',
 '출시',
 '인구',
 '정보처리기사',
 '소프트웨어',
 '설계',
 '프로그래밍',
 '적용',
 '서비스',
 '디자인',
 '필요성',
 '전문의',
 '데이터',
 '사업',
 '고객',
 '직종',
 '훈련',
 '축소',
 '작성',
 '지식',
 '안전교육',
 '파악',
 '변화',
 '개선',
 '침체',
 '정비',
 '자동',
 '항공기',
 '기계',
 '건축',
 '판단',
 '정보',
 '교체',
 '디지털',
 '직접',
 '편집',
 '기계로',
 '영상',
 '치료',
 '프로그램',
 '인공',
 '업체',
 '전기',
 '교육',
 '향상',
 '분석',
 '구조조정',
 '수가',
 '직무교육',
 '개인',
 '문화',
 '스트레스',
 '미디어',
 '실무교육',
 '처리',
 '시장',
 '로봇',
 '신규',
 '검사',
 '위한',
 '효율',
 '건설',
 '조사',
 '병원',
 '산업기사',
 '수준',
 '면허증',
 '빅데이터',
 '상담',
 '영역',
 '지능',
 '관리자',
 '판매',
 '재교육',
 '자격증',
 '의사',
 '사회복지사',
 '시스템',
 '구조',
 '기업',
 '운전면허증',
 '기본',
 '회사',
 '인원',
 '환자',
 '전산',
 '제작',
 '화로',
 '감축',
 '매체',
 '온라인',
 '연구',
 '드론',
 '시설',
 '영향',
 '경기',
 '의료',
 '활성화',
 '새로운',
 '고령화',
 '현장',
 '설비',
 '방법',
 '인터넷',
 '관리',
 '통신',
 '불황',
 '방송',
 '기사',
 '지속',
 '의사면허증',
 '전기기사',
 '가능',
 '요구',
 '토목기사',
 '건강',
 '제조업',
 '학생',
 '개발',
 '자동차',
 '기계화',
 '경제',
 '마케팅',
 '전문',
 '예술',
 '수입',
 '운영',
 '장비',
 '기술',
 '컴퓨터',
 '무량',
 '기능사',
 '자동화

In [25]:
clf2019 = setup(know_train[2], target = 'knowcode',
            fold=5,
            numeric_features = numeric_cols_2019 + numeric_pure_cols_2019 + binary_cols_2019 + half_cols_2019,
            categorical_features = categorical_cols_2019 + korean_cols_2019,
            ignore_features = ['idx', 'description'],
            session_id=RANDOM_STATE,
            silent=True,
            #feature_selection=True,
            #feature_selection_threshold=0.5,
           )

In [26]:
et_2019 = create_model('et')

In [27]:
#evaluate_model(et_2019)

In [27]:
fi_2019 = pd.DataFrame({'Feature': get_config('X_train').columns, 'Value' : abs(et_2019.feature_importances_)}).sort_values(by='Value', ascending=False)

In [28]:
fi_2019

Feature     Value
82                         bq3  0.008363
129                       bq26  0.008066
134                     bq31_2  0.008028
87                         bq7  0.007824
132                       bq30  0.007782
...                        ...       ...
1466   able_job_기계 장비 구매.판매 업무  0.000000
1498          able_job_낙농업 컨설팅  0.000000
1503        able_job_노인요양원 관리자  0.000000
2654               major_바텐더학과  0.000000
1419  able_job_관련 기상사업자  컨설팅업체  0.000000

[3693 rows x 2 columns]

In [29]:
fi_2019.to_csv('fi_2019.csv',encoding='utf-8', index=False)

### 2020

In [7]:
numeric_cols_2020 = ['saq1_1', 'saq1_2', 'saq2_1', 'saq2_2', 'saq3_1', 'saq3_2', 'saq4_1', 'saq4_2', 'saq5_1', 'saq5_2',
                     'saq6_1', 'saq6_2', 'saq7_1', 'saq7_2', 'saq8_1', 'saq8_2', 'saq9_1', 'saq9_2', 'saq10_1', 'saq10_2',
                    'saq11_1', 'saq11_2', 'saq12_1', 'saq12_2', 'saq13_1', 'saq13_2', 'saq14_1', 'saq14_2', 'saq15_1',
                     'saq15_2', 'saq16_1', 'saq16_2', 'saq17_1', 'saq17_2', 'saq18_1', 'saq18_2', 'saq19_1', 'saq19_2',
                     'saq20_1', 'saq20_2','saq21_1', 'saq21_2', 'saq22_1', 'saq22_2', 'saq23_1', 'saq23_2', 'saq24_1', 'saq24_2',
                     'saq25_1', 'saq25_2', 'saq26_1', 'saq26_2', 'saq27_1', 'saq27_2', 'saq28_1', 'saq28_2', 'saq29_1', 'saq29_2',
                     'saq30_1', 'saq30_2', 'saq31_1', 'saq31_2', 'saq32_1', 'saq32_2', 'saq33_1', 'saq33_2', 'saq34_1',
                     'saq34_2', 'saq35_1', 'saq35_2', 'saq36_1', 'saq36_2', 'saq37_1', 'saq37_2', 'saq38_1', 'saq38_2',
                     'saq39_1', 'saq39_2', 'saq40_1', 'saq40_2', 'saq41_1', 'saq41_2', 'saq42_1', 'saq42_2', 'saq43_1',
                     'saq43_2', 'saq44_1', 'saq44_2',
                    'vq1', 'vq2', 'vq3', 'vq4', 'vq5', 'vq6', 'vq7', 'vq8', 'vq9', 'vq10', 'vq11', 'vq12', 'vq13',
                    'bq3', 'bq5_1', 'bq7', 'bq8_1', 'bq8_2', 'bq8_3', 'bq9', 'bq10', 'bq11', 'bq12_1', 'bq12_5', 
                     'bq13_1', 'bq13_2', 'bq13_3', 'bq14_1', 'bq14_2', 'bq14_3', 'bq14_4', 'bq14_5', 'bq14_6', 'bq14_7', 
                     'bq15', 'bq16_1', 'bq16_2', 'bq16_3', 'bq16_4', 'bq16_5', 'bq16_6', 'bq16_7', 'bq16_8', 'bq16_9', 
                     'bq16_10', 'bq17', 'bq18_1', 'bq18_2', 'bq18_3', 'bq18_4', 'bq18_5', 'bq18_6', 'bq18_7', 'bq18_8', 
                     'bq18_9', 'bq19', 'bq20', 'bq21_1', 'bq21_2', 'bq21_3', 'bq22_1', 'bq22_2', 'bq22_3', 
                     'bq22_4', 'bq22_5', 'bq22_6', 'bq23_1', 'bq23_2', 'bq23_3', 'bq26',]
numeric_pure_cols_2020 = ['bq25', 'bq29', 'bq30_1', 'bq30_2', 'bq30_3', 'ubda_cnt']
categorical_cols_2020 = ['bq1', 'bq2', 'bq27_1', 'bq27_2','bq6_mark', 'bq12_2_mark', 'bq12_3_mark', 'bq12_4_mark',]
binary_cols_2020 = ['bq4', 'bq5', 'bq21_4', 'bq24', 'bq28',]
half_cols_2020 = ['bq6', 'bq12_2', 'bq12_3', 'bq12_4', ]
text_cols_2020 = ['bq4_1a', 'bq4_1b', 'bq4_1c', 'bq5_2', 'bq18_10', 'bq20_1', 'bq26_1',]

In [8]:
all_cols_2020 = numeric_cols_2020 + numeric_pure_cols_2020\
                + categorical_cols_2020 + binary_cols_2020 + half_cols_2020 + text_cols_2020

In [9]:
all_nontext_cols_2020 = numeric_cols_2020 + numeric_pure_cols_2020\
                + categorical_cols_2020 + binary_cols_2020 + half_cols_2020

In [10]:
no_match_cols_2020 = [col for col in know_train[3].columns if col not in all_cols_2020]
no_match_cols_2020

['idx',
 'major',
 'description',
 '제품',
 '기기',
 '인구',
 '상승',
 '첨단',
 '정보처리기사',
 '전기산업기사',
 '소프트웨어',
 '불가',
 '서비스',
 '디자인',
 '필요성',
 '전문의',
 '게임',
 '공연',
 '사업',
 '고객',
 '운전면허',
 '직종',
 '훈련',
 '예정',
 '축소',
 '방식',
 '레지던트',
 '안전교육',
 '모바일',
 '변화',
 '기술사',
 '전기기능사',
 '업그레이드',
 '침체',
 '정비',
 '자동',
 '기계',
 '건축',
 '에너지',
 '정보',
 '수리',
 '직접',
 '업계',
 '프로그램',
 '인공',
 '업체',
 '2급',
 '전기',
 '교육',
 '향상',
 '분석',
 '구조조정',
 '수가',
 '직무교육',
 '문화',
 '스트레스',
 '실무교육',
 '변동',
 '비대',
 '시장',
 '로봇',
 '신규',
 '건설',
 '병원',
 '산업기사',
 '면허증',
 '향후',
 '기피',
 '상담',
 '보험',
 '지능',
 '판매',
 '사회복지사',
 '자격증',
 '의사',
 '출산율',
 '시스템',
 '세계',
 '구조',
 '기업',
 '안정',
 '경쟁',
 '인원',
 '환자',
 '전산',
 '성장',
 '화로',
 '감축',
 '채용',
 '가능성',
 '온라인',
 '연구',
 '저하',
 '콘텐츠',
 '무인',
 '시설',
 '화공기사',
 '생활',
 '코로나',
 '영향',
 '박사학위',
 '경기',
 '의료',
 '활성화',
 '요인',
 '고령화',
 '현장',
 '소비',
 '1급',
 '설비',
 '방법',
 '인터넷',
 '관리',
 '통신',
 '불황',
 '방송',
 '기사',
 '지속',
 '조정',
 '전기기사',
 '실습',
 '가능',
 '요구',
 '건강',
 '기능',
 '학생',
 '개발',
 '면허',
 '자동차',
 '기계화',
 '경제',
 '전문',

In [11]:
korean_cols_2020 = [col for col in no_match_cols_2020 if not col in ['idx','major','description','knowcode']] 

In [12]:
korean_cols_2020

['제품',
 '기기',
 '인구',
 '상승',
 '첨단',
 '정보처리기사',
 '전기산업기사',
 '소프트웨어',
 '불가',
 '서비스',
 '디자인',
 '필요성',
 '전문의',
 '게임',
 '공연',
 '사업',
 '고객',
 '운전면허',
 '직종',
 '훈련',
 '예정',
 '축소',
 '방식',
 '레지던트',
 '안전교육',
 '모바일',
 '변화',
 '기술사',
 '전기기능사',
 '업그레이드',
 '침체',
 '정비',
 '자동',
 '기계',
 '건축',
 '에너지',
 '정보',
 '수리',
 '직접',
 '업계',
 '프로그램',
 '인공',
 '업체',
 '2급',
 '전기',
 '교육',
 '향상',
 '분석',
 '구조조정',
 '수가',
 '직무교육',
 '문화',
 '스트레스',
 '실무교육',
 '변동',
 '비대',
 '시장',
 '로봇',
 '신규',
 '건설',
 '병원',
 '산업기사',
 '면허증',
 '향후',
 '기피',
 '상담',
 '보험',
 '지능',
 '판매',
 '사회복지사',
 '자격증',
 '의사',
 '출산율',
 '시스템',
 '세계',
 '구조',
 '기업',
 '안정',
 '경쟁',
 '인원',
 '환자',
 '전산',
 '성장',
 '화로',
 '감축',
 '채용',
 '가능성',
 '온라인',
 '연구',
 '저하',
 '콘텐츠',
 '무인',
 '시설',
 '화공기사',
 '생활',
 '코로나',
 '영향',
 '박사학위',
 '경기',
 '의료',
 '활성화',
 '요인',
 '고령화',
 '현장',
 '소비',
 '1급',
 '설비',
 '방법',
 '인터넷',
 '관리',
 '통신',
 '불황',
 '방송',
 '기사',
 '지속',
 '조정',
 '전기기사',
 '실습',
 '가능',
 '요구',
 '건강',
 '기능',
 '학생',
 '개발',
 '면허',
 '자동차',
 '기계화',
 '경제',
 '전문',
 '수입',
 '기술',
 '장비',
 '컴퓨터',
 '무량

In [13]:
clf2020 = setup(know_train[3], target = 'knowcode',
            fold=5,
            numeric_features = numeric_cols_2020 + numeric_pure_cols_2020 + binary_cols_2020 + half_cols_2020,
            categorical_features = categorical_cols_2020 + korean_cols_2020,
            ignore_features = ['idx', 'description'],
            session_id=RANDOM_STATE,
            silent=True,
            #feature_selection=True,
            #feature_selection_threshold=0.5,
           )

In [14]:
et_2020 = create_model('et')

In [15]:
#evaluate_model(et_2020)

In [16]:
fi_2020 = pd.DataFrame({'Feature': get_config('X_train').columns, 'Value' : abs(et_2020.feature_importances_)}).sort_values(by='Value', ascending=False)

In [17]:
fi_2020

Feature     Value
170         bq30_2  0.007186
106            bq7  0.007130
101            bq3  0.006464
168           bq29  0.006262
105            bq6  0.006096
...            ...       ...
657      major_안보학  0.000000
662    major_애완동물학  0.000000
1008    major_표면처리  0.000000
1004  major_특수산업설비  0.000000
307     major_국제관광  0.000000

[1234 rows x 2 columns]

In [18]:
fi_2020.to_csv('fi_2020.csv',encoding='utf-8', index=False)

## Option 1: bq 문항과 text 문항만 사용 

In [11]:
use_cols_option1_2017 = ['bq1','bq2','bq3','bq4','bq5','bq5_1','bq6','bq7','bq8_1', 'bq8_2', 'bq8_3',
 'bq9','bq10','bq11','bq12_1','bq12_2','bq12_3','bq12_4','bq12_5','bq13','bq14','bq15_1','bq15_2',
 'bq15_3','bq16','bq17','bq18_1','bq18_2','bq18_3','bq18_4','bq18_5','bq18_6','bq18_7','bq19','bq20',
 'bq21','bq22','bq23','bq24_1','bq24_2','bq24_3','bq24_4','bq24_5','bq24_6','bq24_7','bq24_8','bq25',
 'bq26','bq27','bq28','bq29','bq35','bq36','bq37','bq38','bq39_1','bq39_2','bq40','bq41_1','bq41_2',
 'bq41_3','sim_job','bef_job','able_job','major','description','bq6_mark','bq12_2_mark','bq12_3_mark','bq12_4_mark',
 '제품','기기','인구','줄자','공급','정보처리기사','필기','오토캐드','포화상태','설계','음악',
 '회계','카메라','서비스','디자인','전문의','사업','무대','고객','직종','훈련','축소','지식','안전교육',
 '파악','변화','드릴','침체','자동','기계','현미경','건축','장갑','에너지','정보','캐드','기록','정부',
 '무전기','편집','가위','기구','마이크','의상','치료','프로그램','인공','업체','전기','교육','향상','보안',
 '분석','수가','드라이버','직무교육','문화','자료','실무교육','변동','인식','시장','로봇','검사','전화기',
 '노인','엑셀','건설','병원','면허증','향후','기피','노트북','포토샵','단기','보험','지능','판매','피아노','자격증',
 '볼펜','시스템','기업','운전면허증','회사','인원','환자','전산','오피스','화로','망치','복지','감축','간호사','활동',
 '매체','시험','연구','교원자격증','한글','시설','생활','교재','영향','특성','측정','정책','경기','의료',
 '공구','활성화','고령화','배출','설비','방법','인터넷','관리','실험','측정기','불황','방송','지속','의사면허증',
 '전기기사','건강','제조업','학생','워드','개발','대본','접기','통계','자동차','기계화','일러스트','현상','도면',
 '전문','파워포인트','예술','수입','장비','기술','컴퓨터','청진기','자동화','생산','안전','계측기','해외','작업',
 'ubda_cnt','knowcode']

In [10]:
list(know_train[0].columns)

['idx',
 'aq1_1',
 'aq1_2',
 'aq2_1',
 'aq2_2',
 'aq3_1',
 'aq3_2',
 'aq4_1',
 'aq4_2',
 'aq5_1',
 'aq5_2',
 'aq6_1',
 'aq6_2',
 'aq7_1',
 'aq7_2',
 'aq8_1',
 'aq8_2',
 'aq9_1',
 'aq9_2',
 'aq10_1',
 'aq10_2',
 'aq11_1',
 'aq11_2',
 'aq12_1',
 'aq12_2',
 'aq13_1',
 'aq13_2',
 'aq14_1',
 'aq14_2',
 'aq15_1',
 'aq15_2',
 'aq16_1',
 'aq16_2',
 'aq17_1',
 'aq17_2',
 'aq18_1',
 'aq18_2',
 'aq19_1',
 'aq19_2',
 'aq20_1',
 'aq20_2',
 'aq21_1',
 'aq21_2',
 'aq22_1',
 'aq22_2',
 'aq23_1',
 'aq23_2',
 'aq24_1',
 'aq24_2',
 'aq25_1',
 'aq25_2',
 'aq26_1',
 'aq26_2',
 'aq27_1',
 'aq27_2',
 'aq28_1',
 'aq28_2',
 'aq29_1',
 'aq29_2',
 'aq30_1',
 'aq30_2',
 'aq31_1',
 'aq31_2',
 'aq32_1',
 'aq32_2',
 'aq33_1',
 'aq33_2',
 'aq34_1',
 'aq34_2',
 'aq35_1',
 'aq35_2',
 'aq36_1',
 'aq36_2',
 'aq37_1',
 'aq37_2',
 'aq38_1',
 'aq38_2',
 'aq39_1',
 'aq39_2',
 'aq40_1',
 'aq40_2',
 'aq41_1',
 'aq41_2',
 'bq1',
 'bq2',
 'bq3',
 'bq4',
 'bq5',
 'bq5_1',
 'bq6',
 'bq7',
 'bq8_1',
 'bq8_2',
 'bq8_3',
 'bq9',
 'bq

In [9]:
know_train[0].columns

Index(['idx', 'aq1_1', 'aq1_2', 'aq2_1', 'aq2_2', 'aq3_1', 'aq3_2', 'aq4_1',
       'aq4_2', 'aq5_1',
       ...
       'bq41_1', 'bq41_2', 'bq41_3', 'knowcode', 'text_response', 'sim_job',
       'bef_job', 'able_job', 'major', 'description'],
      dtype='object', length=151)

In [10]:
know_train[1].columns 

Index(['idx', 'cq1', 'cq2', 'cq3', 'cq4', 'cq5', 'cq6', 'cq7', 'cq8', 'cq9',
       ...
       'bq41_1', 'bq41_2', 'bq41_3', 'knowcode', 'text_response', 'sim_job',
       'bef_job', 'able_job', 'major', 'description'],
      dtype='object', length=136)

In [11]:
know_train[2].columns 

Index(['idx', 'sq1', 'sq2', 'sq3', 'sq4', 'sq5', 'sq6', 'sq7', 'sq8', 'sq9',
       ...
       'bq30', 'bq31_1', 'bq31_2', 'bq31_3', 'knowcode', 'text_response',
       'bef_job', 'able_job', 'major', 'description'],
      dtype='object', length=148)

In [12]:
know_train[3].columns 

Index(['idx', 'saq1_1', 'saq1_2', 'saq2_1', 'saq2_2', 'saq3_1', 'saq3_2',
       'saq4_1', 'saq4_2', 'saq5_1',
       ...
       'bq27_2', 'bq28', 'bq29', 'bq30_1', 'bq30_2', 'bq30_3', 'knowcode',
       'text_response', 'major', 'description'],
      dtype='object', length=181)

2017: sim_job, bef_job, able_job, major<br/>
2018: sim_job, bef_job, able_job, major<br/>
2019: bef_job, able_job, major<br/>
2020: major<br/>

# 모든 문항 분류 체크

### 2017

In [13]:
numeric_cols_2017 = ['aq1_1', 'aq1_2', 'aq2_1', 'aq2_2', 'aq3_1', 'aq3_2', 'aq4_1', 'aq4_2', 'aq5_1', 'aq5_2', 
                     'aq6_1', 'aq6_2', 'aq7_1', 'aq7_2', 'aq8_1', 'aq8_2', 'aq9_1', 'aq9_2', 'aq10_1', 'aq10_2',
                    'aq11_1', 'aq11_2', 'aq12_1', 'aq12_2', 'aq13_1', 'aq13_2', 'aq14_1', 'aq14_2', 'aq15_1', 'aq15_2', 
                     'aq16_1', 'aq16_2', 'aq17_1', 'aq17_2', 'aq18_1', 'aq18_2', 'aq19_1', 'aq19_2', 'aq20_1', 'aq20_2',
                    'aq21_1', 'aq21_2', 'aq22_1', 'aq22_2', 'aq23_1', 'aq23_2', 'aq24_1', 'aq24_2', 'aq25_1', 'aq25_2', 
                     'aq26_1', 'aq26_2', 'aq27_1', 'aq27_2', 'aq28_1', 'aq28_2', 'aq29_1', 'aq29_2', 'aq30_1', 'aq30_2',
                    'aq31_1', 'aq31_2', 'aq32_1', 'aq32_2', 'aq33_1', 'aq33_2', 'aq34_1', 'aq34_2', 'aq35_1', 'aq35_2', 
                     'aq36_1', 'aq36_2', 'aq37_1', 'aq37_2', 'aq38_1', 'aq38_2', 'aq39_1', 'aq39_2', 'aq40_1', 'aq40_2',
                     'aq41_1', 'aq41_2', 'bq3', 'bq5_1', 'bq7', 'bq8_1', 'bq8_2', 'bq8_3', 'bq9', 'bq10',
                     'bq11', 'bq12_1', 'bq12_5', 'bq13', 'bq14', 'bq15_1', 'bq15_2', 'bq15_3', 'bq16', 'bq17', 
                     'bq18_1', 'bq18_2', 'bq18_3', 'bq18_4', 'bq18_5', 'bq18_6', 'bq18_7', 'bq19', 'bq20', 'bq21', 
                     'bq22', 'bq25', 'bq26', 'bq27', 'bq28', 'bq29', 'bq35', 'bq37', 'bq38', ]
numeric_pure_cols_2017 = ['bq23', 'bq37', 'bq41_1', 'bq41_2', 'bq41_3',]
categorical_cols_2017 = ['bq1', 'bq2', 'bq39_1', 'bq39_2', ]
binary_cols_2017 = ['bq4', 'bq5', 'bq24_1', 'bq24_2', 'bq24_3', 'bq24_4', 'bq24_5', 'bq24_6', 'bq24_7', 'bq24_8', 'bq36', 'bq40', ]
half_cols_2017 = ['bq6', 'bq12_2', 'bq12_3', 'bq12_4', ]
text_cols_2017 = ['bq4_1a', 'bq4_1b', 'bq4_1c', 'bq5_2', 'bq19_1', 'bq30', 'bq31', 'bq32', 'bq33', 'bq34', 'bq38_1',]

In [14]:
all_cols_2017 = numeric_cols_2017 + numeric_pure_cols_2017\
                + categorical_cols_2017 + binary_cols_2017 + half_cols_2017 + text_cols_2017

In [15]:
all_nontext_cols_2017 = numeric_cols_2017 + numeric_pure_cols_2017\
                + categorical_cols_2017 + binary_cols_2017 + half_cols_2017

all_cols_2017에 없는 열

In [16]:
no_match_cols_2017 = [col for col in know_train[0].columns if col not in all_cols_2017]
no_match_cols_2017

['idx',
 'knowcode',
 'text_response',
 'sim_job',
 'bef_job',
 'able_job',
 'major',
 'description']

In [17]:
know_train[0][[col for col in know_train[0].columns if col not in all_cols_2017]]

idx  knowcode  ...   major    description
0        0    825101  ...      실업    도장원(도장기조작원)
1        1    140204  ...    건축공학       건축설비 기술자
2        2    140204  ...    건축공학       건축설비 기술자
3        3    140601  ...    환경학과     지적 및 측량기술자
4        4    140204  ...    건축공학       건축설비 기술자
...    ...       ...  ...     ...            ...
9481  9481    411301  ...   산업디자인         출판물기획자
9482  9482    151105  ...  우주항공공학        항공공학기술자
9483  9483    701101  ...      기계  철골공(강구조물 건립원)
9484  9484     25402  ...   문헌정보학          법원공무원
9485  9485     15201  ...      전기         운송 관리자

[9486 rows x 8 columns]

all_nontext_cols_2017의 열들은 다 2017에 있는가?

In [18]:
know_train[0][all_nontext_cols_2017]

aq1_1  aq1_2  aq2_1  aq2_2  aq3_1  ...  bq40  bq6  bq12_2  bq12_3  bq12_4
0         3      3      3      3      3  ...     1    5       3       4       4
1         4      5      4      5      3  ...     1    6       3       4       4
2         3      4      3      4      3  ...     1    6       4       4       4
3         3      3      3      3      3  ...     1    2       3       3       3
4         4      5      3      4      3  ...     1    6       3       4       4
...     ...    ...    ...    ...    ...  ...   ...  ...     ...     ...     ...
9481      3      5      2      4      3  ...     1    7       4       4       4
9482      5      5      5      5      5  ...     1    6       3       4       3
9483      3      3      4      6      3  ...     2    6       0       0       4
9484      3      5      3      5      4  ...     1    5       3       4       4
9485      3      4      3      4      3  ...     1    7       5       5       5

[9486 rows x 144 columns]

In [19]:
no_match_cols_2017

['idx',
 'knowcode',
 'text_response',
 'sim_job',
 'bef_job',
 'able_job',
 'major',
 'description']

### half_cols 전처리

In [20]:
for col in half_cols_2017:
    print(col)
    print(know_train[0][col].value_counts())

bq6
7    2074
6    1919
5    1448
3    1365
4    1106
2    1053
1     521
Name: bq6, dtype: int64
bq12_2
3    4194
4    2088
0    1630
2    1113
5     284
1     177
Name: bq12_2, dtype: int64
bq12_3
3    3641
4    3371
0    1343
5     556
2     515
1      60
Name: bq12_3, dtype: int64
bq12_4
4    4603
3    2985
0     879
5     834
2     170
1      15
Name: bq12_4, dtype: int64


In [21]:
know_train[0]['bq6_mark']= 0
know_train[0]['bq6_mark'].loc[know_train[0]['bq6'] == 7] = 1
know_train[0]['bq12_2_mark']= 0
know_train[0]['bq12_2_mark'].loc[know_train[0]['bq12_2'] == 9] = 1
know_train[0]['bq12_3_mark']= 0
know_train[0]['bq12_3_mark'].loc[know_train[0]['bq12_3'] == 9] = 1
know_train[0]['bq12_4_mark']= 0
know_train[0]['bq12_4_mark'].loc[know_train[0]['bq12_4'] == 9] = 1

In [22]:
know_train[0].loc[know_train[0]['bq6'] == 7, 'bq6'] = 3 # median filling
know_train[0].loc[know_train[0]['bq12_2'] == 9, 'bq12_2'] = 3 # median filling
know_train[0].loc[know_train[0]['bq12_3'] == 9, 'bq12_3'] = 3 # median filling
know_train[0].loc[know_train[0]['bq12_4'] == 9, 'bq12_4'] = 3 # median filling

In [23]:
half_cols_mark_2017 = ['bq6_mark', 'bq12_2_mark', 'bq12_3_mark', 'bq12_4_mark',]

### 2018

전처리: categorical column이어야 하는 bq22, bq23, bq24가 제멋대로 bq231, bq232.. 로 표기되어 있음... 원핫인코딩도 아니고 레이블인코딩도 아닌 무언가...

In [24]:
know_train[1][['bq221',
              'bq222',
              'bq223',
              'bq231',
              'bq232',
              'bq233',
              'bq234',
              'bq235',
              'bq241',
              'bq242',
              'bq243',
              'bq244',
              'bq245',]]

bq221  bq222  bq223  bq231  bq232  ...  bq241  bq242  bq243  bq244  bq245
0         1      0    0.0      1      0  ...      1      0      0      0      0
1         1      0    0.0      1      0  ...      0      0      3      0      0
2         0      2    0.0      0      0  ...      0      0      0      0      5
3         1      0    0.0      1      0  ...      1      0      0      0      0
4         1      0    0.0      1      0  ...      1      0      0      0      0
...     ...    ...    ...    ...    ...  ...    ...    ...    ...    ...    ...
9067      0      2    0.0      0      0  ...      0      0      0      0      5
9068      0      2    0.0      0      0  ...      0      0      0      0      5
9069      0      2    0.0      0      0  ...      0      0      0      0      5
9070      0      2    0.0      0      0  ...      0      0      0      0      5
9071      1      0    0.0      1      0  ...      1      0      0      0      0

[9072 rows x 13 columns]

In [25]:
know_train[1]['bq22'] =  know_train[1]['bq221'] + know_train[1]['bq222'] + know_train[1]['bq223']

In [26]:
know_train[1]['bq23'] =  know_train[1]['bq231'] + know_train[1]['bq232'] + know_train[1]['bq233']+\
                         know_train[1]['bq234'] + know_train[1]['bq235']

In [27]:
know_train[1]['bq24'] =  know_train[1]['bq241'] + know_train[1]['bq242'] + know_train[1]['bq243']+\
                         know_train[1]['bq244'] + know_train[1]['bq245']

In [28]:
know_train[1].drop(['bq221',
              'bq222',
              'bq223',
              'bq231',
              'bq232',
              'bq233',
              'bq234',
              'bq235',
              'bq241',
              'bq242',
              'bq243',
              'bq244',
              'bq245',], axis=1, inplace=True)

In [29]:
numeric_cols_2018 = ['cq1', 'cq2', 'cq3', 'cq4', 'cq5', 'cq6', 'cq7', 'cq8', 'cq9', 'cq10', 
                    'cq11', 'cq12', 'cq13', 'cq14', 'cq15', 'cq16', 'cq17', 'cq18', 'cq19', 'cq20', 
                    'cq21', 'cq22', 'cq23', 'cq24', 'cq25', 'cq26', 'cq27', 'cq28', 'cq29', 'cq30',
                    'cq31', 'cq32', 'cq33', 'cq34', 'cq35', 'cq36', 'cq37', 'cq38', 'cq39', 'cq40', 
                    'cq41', 'cq42', 'cq43', 'cq44', 'cq45', 'cq46', 'cq47', 'cq49', 'cq50_1', 'cq50_2', 'cq50_3', 'cq50_4', 'cq50_5', 'cq50_6', 'cq50_7', 'cq50_8',
                    'iq1', 'iq2', 'iq3', 'iq4', 'iq5', 'iq6',
                    'bq3', 'bq5_1', 'bq7', 'bq8_1', 'bq8_2', 'bq8_3', 'bq9', 'bq10', 'bq11', 'bq12_1', 'bq12_5', 'bq18', 'bq19', 'bq20', 'bq25', 'bq26_1a', 
                     'bq26_2a', 'bq26_3a', 'bq26_4a', 'bq28', 'bq34', 'bq37', ]
numeric_pure_cols_2018 = ['bq21', 'bq36', 'bq40', 'bq41_1', 'bq41_2', 'bq41_3', ]
categorical_cols_2018 = ['bq1', 'bq2', 'bq13', 'bq15', 'bq17', 'bq22', 'bq23', 'bq24', 'bq26_1', 'bq26_2', 'bq26_3', 'bq26_4', 'bq38', 'bq38_1', 'bq38_2', ]
binary_cols_2018 = ['cq48', 'bq4', 'bq5', 'bq14', 'bq16', 'bq25_1', 'bq27', 'bq35', 'bq39', ]
half_cols_2018 = ['bq6', 'bq12_2', 'bq12_3', 'bq12_4', ]
text_cols_2018 = ['bq4_1a', 'bq4_1b', 'bq4_1c', 'bq5_2', 'bq28_1', 'bq29', 'bq30', 'bq31', 'bq32', 'bq33', 'bq37_1', ]

In [30]:
all_cols_2018 = numeric_cols_2018 + numeric_pure_cols_2018\
                + categorical_cols_2018 + binary_cols_2018 + half_cols_2018 + text_cols_2018

In [31]:
all_nontext_cols_2018 = numeric_cols_2018 + numeric_pure_cols_2018\
                + categorical_cols_2018 + binary_cols_2018 + half_cols_2018

all_cols_2018에 없는 열

In [32]:
no_match_cols_2018 = [col for col in know_train[1].columns if col not in all_cols_2018]
no_match_cols_2018

['idx',
 'knowcode',
 'text_response',
 'sim_job',
 'bef_job',
 'able_job',
 'major',
 'description']

In [33]:
know_train[1][[col for col in know_train[1].columns if col not in all_cols_2018]]

idx  knowcode  ...  major                                description
0      9486    306301  ...   치기공과                                치과기공사 치과기공사
1      9487    133204  ...     이과                           게임프로그래머 게임 프로그래머
2      9488    702502  ...    기계과                    유리 부착원(유리공) 유리 부착원(유리공)
3      9489    121201  ...  물리 과학                          자연과학 시험원 자연과학 시험원
4      9490    306301  ...   치기공학                                치과기공사 치과기공사
...     ...       ...  ...    ...                                        ...
9067  18564    303002  ...    한약학                                    한약사 한약사
9068  18565    615203  ...    국문과                              통신 기기·서비스 판매원
9069  18566    301106  ...     의대                                  안과의사 안과의사
9070  18567    301112  ...     의예                                피부과의사 피부과의사
9071  18568    133101  ...  컴퓨터공학  시스템 소프트웨어 개발자(프로그래머) 시스템 소프트웨어 개발자(프로그래머)

[9072 rows x 8 columns]

all_nontext_cols_2018의 열들은 다 2018에 있는가?

In [34]:
know_train[1][all_nontext_cols_2018]

cq1  cq2  cq3  cq4  cq5  cq6  ...  bq35  bq39  bq6  bq12_2  bq12_3  bq12_4
0       5    3    4    2    2    5  ...   1.0     1    7       4       4       4
1       4    3    4    4    4    4  ...   1.0     1    7       0       5       5
2       4    1    5    3    3    4  ...   1.0     0    5       2       2       0
3       4    3    4    4    4    3  ...   1.0     1    4       4       4       4
4       3    2    3    1    1    4  ...   2.0     1    7       0       4       4
...   ...  ...  ...  ...  ...  ...  ...   ...   ...  ...     ...     ...     ...
9067    5    5    5    2    2    4  ...   1.0     0    3       0       0       0
9068    5    5    5    3    3    5  ...   1.0     0    2       0       0       0
9069    4    4    4    3    3    5  ...   2.0     0    6       0       0       4
9070    5    2    3    2    3    5  ...   1.0     0    6       0       0       3
9071    5    4    5    4    3    5  ...   1.0     1    7       4       4       4

[9072 rows x 118 columns]

In [35]:
no_match_cols_2018

['idx',
 'knowcode',
 'text_response',
 'sim_job',
 'bef_job',
 'able_job',
 'major',
 'description']

### half_cols 전처리

In [36]:
for col in half_cols_2018:
    print(col)
    print(know_train[0][col].value_counts())

bq6
3    3439
6    1919
5    1448
4    1106
2    1053
1     521
Name: bq6, dtype: int64
bq12_2
3    4194
4    2088
0    1630
2    1113
5     284
1     177
Name: bq12_2, dtype: int64
bq12_3
3    3641
4    3371
0    1343
5     556
2     515
1      60
Name: bq12_3, dtype: int64
bq12_4
4    4603
3    2985
0     879
5     834
2     170
1      15
Name: bq12_4, dtype: int64


In [37]:
know_train[1]['bq6_mark']= 0
know_train[1]['bq6_mark'].loc[know_train[1]['bq6'] == 7] = 1
know_train[1]['bq12_2_mark']= 0
know_train[1]['bq12_2_mark'].loc[know_train[1]['bq12_2'] == 9] = 1
know_train[1]['bq12_3_mark']= 0
know_train[1]['bq12_3_mark'].loc[know_train[1]['bq12_3'] == 9] = 1
know_train[1]['bq12_4_mark']= 0
know_train[1]['bq12_4_mark'].loc[know_train[1]['bq12_4'] == 9] = 1

In [38]:
know_train[1].loc[know_train[1]['bq6'] == 7, 'bq6'] = 3 # median filling
know_train[1].loc[know_train[1]['bq12_2'] == 9, 'bq12_2'] = 3 # median filling
know_train[1].loc[know_train[1]['bq12_3'] == 9, 'bq12_3'] = 3 # median filling
know_train[1].loc[know_train[1]['bq12_4'] == 9, 'bq12_4'] = 3 # median filling

In [39]:
half_cols_mark_2018 = ['bq6_mark', 'bq12_2_mark', 'bq12_3_mark', 'bq12_4_mark',]

### 2019

In [40]:
numeric_cols_2019 = ['sq1', 'sq2', 'sq3', 'sq4', 'sq5', 'sq6', 'sq7', 'sq8', 'sq9', 'sq10', 
                     'sq11', 'sq12', 'sq13', 'sq14', 'sq15', 'sq16',
                    'kq1_1', 'kq1_2', 'kq2_1', 'kq2_2', 'kq3_1', 'kq3_2', 'kq4_1', 'kq4_2', 'kq5_1', 'kq5_2', 
                     'kq6_1', 'kq6_2', 'kq7_1', 'kq7_2', 'kq8_1', 'kq8_2', 'kq9_1', 'kq9_2', 'kq10_1', 'kq10_2',
                    'kq11_1', 'kq11_2', 'kq12_1', 'kq12_2', 'kq13_1', 'kq13_2', 'kq14_1', 'kq14_2', 'kq15_1', 'kq15_2',
                     'kq16_1', 'kq16_2', 'kq17_1', 'kq17_2', 'kq18_1', 'kq18_2', 'kq19_1', 'kq19_2', 'kq20_1', 'kq20_2',
                    'kq21_1', 'kq21_2', 'kq22_1', 'kq22_2', 'kq23_1', 'kq23_2', 'kq24_1', 'kq24_2', 'kq25_1', 'kq25_2', 
                     'kq26_1', 'kq26_2', 'kq27_1', 'kq27_2', 'kq28_1', 'kq28_2', 'kq29_1', 'kq29_2', 'kq30_1', 'kq30_2',
                    'kq31_1', 'kq31_2', 'kq32_1', 'kq32_2', 'kq33_1', 'kq33_2', 
                    'bq3', 'bq5_1', 'bq7', 'bq8_1', 'bq8_2', 'bq8_3', 'bq9', 'bq10', 'bq11', 'bq12_1', 'bq12_5', 
                     'bq13_1', 'bq13_2', 'bq13_3', 'bq14_1', 'bq14_2', 'bq14_3', 'bq14_4', 'bq14_5', 'bq15', 
                     'bq16_1', 'bq16_2', 'bq16_3', 'bq16_4', 'bq16_5', 'bq17', 'bq19', 'bq20', 'bq21_1', 'bq21_2', 'bq21_3', 'bq27',]
numeric_pure_cols_2019 = ['bq26', 'bq30', 'bq31_1', 'bq31_2', 'bq31_3', ]
categorical_cols_2019 = ['bq1', 'bq2', 'bq28', 'bq28_1', 'bq28_2', ]
binary_cols_2019 = ['bq4', 'bq5', 'bq18_1', 'bq18_2', 'bq18_3', 'bq18_4', 'bq18_5', 'bq18_6', 'bq18_7', 
                    'bq18_8', 'bq18_9', 'bq25', 'bq29', ]
half_cols_2019 = ['bq6', 'bq12_2', 'bq12_3', 'bq12_4', ]
text_cols_2019 = ['bq4_1a', 'bq4_1b', 'bq4_1c', 'bq5_2', 'bq18_10', 'bq20_1', 'bq22', 'bq23', 'bq24', 'bq27_1',]

In [41]:
all_cols_2019 = numeric_cols_2019 + numeric_pure_cols_2019\
                + categorical_cols_2019 + binary_cols_2019 + half_cols_2019 + text_cols_2019

In [42]:
all_nontext_cols_2019 = numeric_cols_2019 + numeric_pure_cols_2019\
                + categorical_cols_2019 + binary_cols_2019 + half_cols_2019

all_cols_2019에 없는 열

In [43]:
no_match_cols_2019 = [col for col in know_train[2].columns if col not in all_cols_2019]
no_match_cols_2019

['idx',
 'knowcode',
 'text_response',
 'bef_job',
 'able_job',
 'major',
 'description']

In [44]:
know_train[2][[col for col in know_train[2].columns if col not in all_cols_2019]]

idx  knowcode  ...   major                                    description
0     18569    812301  ...   기계공학과                    철도기관차·전동차 정비원 철도기관차·전동차 정비원
1     18570     15201  ...     경영학                                  운송 관리자 운송 관리자
2     18571    901101  ...       0                              곡식작물 재배원 곡식작물 재배원
3     18572     29903  ...     행정학                            행정사·문서대행자 행정사·문서대행자
4     18573    561401  ...       0                                         구두 미화원
...     ...       ...  ...     ...                                            ...
8550  27122    416105  ...   방송영상과                          연극영화방송기술감독 연극영화방송기술감독
8551  27123    531601  ...   호텔관광학                              바텐터(조주사) 바텐터(조주사)
8552  27124    615301  ...     상업과                                온라인 판매원 온라인 판매원
8553  27125    811901  ...      기계  농업용 및 기타 기계장비 설치 및 정비원 농업용 및 기타 기계장비 설치 및 정비원
8554  27126    812101  ...  항공정비공학              항공기 정비원(헬리콥터 제외) 항공기 정비원(헬리콥터 제외)

[8555 rows x 7 columns]

all_nontext_cols_2019의 열들은 다 2019에 있는가?

In [45]:
know_train[2][all_nontext_cols_2019]

sq1  sq2  sq3  sq4  sq5  sq6  ...  bq25  bq29  bq6  bq12_2  bq12_3  bq12_4
0       4    4    4    3    4    4  ...   1.0     1  7.0       5       4     4.0
1       4    3    4    4    4    4  ...   1.0     0  1.0       9       9     4.0
2       2    3    2    2    2    2  ...   2.0     0  2.0       9       9     9.0
3       5    5    5    5    5    5  ...   2.0     1  4.0       3       3     4.0
4       1    4    4    1    1    2  ...   1.0     0  7.0       9       9     9.0
...   ...  ...  ...  ...  ...  ...  ...   ...   ...  ...     ...     ...     ...
8550    3    4    4    3    4    3  ...   1.0     1  4.0       3       4     4.0
8551    5    5    4    4    4    5  ...   1.0     1  4.0       3       5     5.0
8552    3    3    4    4    4    4  ...   2.0     1  5.0       3       4     3.0
8553    3    3    3    1    4    4  ...   1.0     1  3.0       3       3     3.0
8554    5    5    5    5    5    3  ...   1.0     1  6.0       3       3     5.0

[8555 rows x 141 columns]

### half_cols 전처리

In [46]:
for col in half_cols_2019:
    print(col)
    print(know_train[0][col].value_counts())

bq6
3    3439
6    1919
5    1448
4    1106
2    1053
1     521
Name: bq6, dtype: int64
bq12_2
3    4194
4    2088
0    1630
2    1113
5     284
1     177
Name: bq12_2, dtype: int64
bq12_3
3    3641
4    3371
0    1343
5     556
2     515
1      60
Name: bq12_3, dtype: int64
bq12_4
4    4603
3    2985
0     879
5     834
2     170
1      15
Name: bq12_4, dtype: int64


In [47]:
know_train[2]['bq6_mark']= 0
know_train[2]['bq6_mark'].loc[know_train[2]['bq6'] == 7] = 1
know_train[2]['bq12_2_mark']= 0
know_train[2]['bq12_2_mark'].loc[know_train[2]['bq12_2'] == 9] = 1
know_train[2]['bq12_3_mark']= 0
know_train[2]['bq12_3_mark'].loc[know_train[2]['bq12_3'] == 9] = 1
know_train[2]['bq12_4_mark']= 0
know_train[2]['bq12_4_mark'].loc[know_train[2]['bq12_4'] == 9] = 1

In [48]:
know_train[2].loc[know_train[2]['bq6'] == 7, 'bq6'] = 3 # median filling
know_train[2].loc[know_train[2]['bq12_2'] == 9, 'bq12_2'] = 3 # median filling
know_train[2].loc[know_train[2]['bq12_3'] == 9, 'bq12_3'] = 3 # median filling
know_train[2].loc[know_train[2]['bq12_4'] == 9, 'bq12_4'] = 3 # median filling

In [49]:
half_cols_mark_2019 = ['bq6_mark', 'bq12_2_mark', 'bq12_3_mark', 'bq12_4_mark',]

### 2020

In [50]:
numeric_cols_2020 = ['saq1_1', 'saq1_2', 'saq2_1', 'saq2_2', 'saq3_1', 'saq3_2', 'saq4_1', 'saq4_2', 'saq5_1', 'saq5_2',
                     'saq6_1', 'saq6_2', 'saq7_1', 'saq7_2', 'saq8_1', 'saq8_2', 'saq9_1', 'saq9_2', 'saq10_1', 'saq10_2',
                    'saq11_1', 'saq11_2', 'saq12_1', 'saq12_2', 'saq13_1', 'saq13_2', 'saq14_1', 'saq14_2', 'saq15_1',
                     'saq15_2', 'saq16_1', 'saq16_2', 'saq17_1', 'saq17_2', 'saq18_1', 'saq18_2', 'saq19_1', 'saq19_2',
                     'saq20_1', 'saq20_2','saq21_1', 'saq21_2', 'saq22_1', 'saq22_2', 'saq23_1', 'saq23_2', 'saq24_1', 'saq24_2',
                     'saq25_1', 'saq25_2', 'saq26_1', 'saq26_2', 'saq27_1', 'saq27_2', 'saq28_1', 'saq28_2', 'saq29_1', 'saq29_2',
                     'saq30_1', 'saq30_2', 'saq31_1', 'saq31_2', 'saq32_1', 'saq32_2', 'saq33_1', 'saq33_2', 'saq34_1',
                     'saq34_2', 'saq35_1', 'saq35_2', 'saq36_1', 'saq36_2', 'saq37_1', 'saq37_2', 'saq38_1', 'saq38_2',
                     'saq39_1', 'saq39_2', 'saq40_1', 'saq40_2', 'saq41_1', 'saq41_2', 'saq42_1', 'saq42_2', 'saq43_1',
                     'saq43_2', 'saq44_1', 'saq44_2',
                    'vq1', 'vq2', 'vq3', 'vq4', 'vq5', 'vq6', 'vq7', 'vq8', 'vq9', 'vq10', 'vq11', 'vq12', 'vq13',
                    'bq3', 'bq5_1', 'bq7', 'bq8_1', 'bq8_2', 'bq8_3', 'bq9', 'bq10', 'bq11', 'bq12_1', 'bq12_5', 
                     'bq13_1', 'bq13_2', 'bq13_3', 'bq14_1', 'bq14_2', 'bq14_3', 'bq14_4', 'bq14_5', 'bq14_6', 'bq14_7', 
                     'bq15', 'bq16_1', 'bq16_2', 'bq16_3', 'bq16_4', 'bq16_5', 'bq16_6', 'bq16_7', 'bq16_8', 'bq16_9', 
                     'bq16_10', 'bq17', 'bq18_1', 'bq18_2', 'bq18_3', 'bq18_4', 'bq18_5', 'bq18_6', 'bq18_7', 'bq18_8', 
                     'bq18_9', 'bq19', 'bq20', 'bq21_1', 'bq21_2', 'bq21_3', 'bq22_1', 'bq22_2', 'bq22_3', 
                     'bq22_4', 'bq22_5', 'bq22_6', 'bq23_1', 'bq23_2', 'bq23_3', 'bq26',]
numeric_pure_cols_2020 = ['bq25', 'bq29', 'bq30_1', 'bq30_2', 'bq30_3', ]
categorical_cols_2020 = ['bq1', 'bq2', 'bq27_1', 'bq27_2',]
binary_cols_2020 = ['bq4', 'bq5', 'bq21_4', 'bq24', 'bq28',]
half_cols_2020 = ['bq6', 'bq12_2', 'bq12_3', 'bq12_4', ]
text_cols_2020 = ['bq4_1a', 'bq4_1b', 'bq4_1c', 'bq5_2', 'bq18_10', 'bq20_1', 'bq26_1',]

In [51]:
all_cols_2020 = numeric_cols_2020 + numeric_pure_cols_2020\
                + categorical_cols_2020 + binary_cols_2020 + half_cols_2020 + text_cols_2020

In [52]:
all_nontext_cols_2020 = numeric_cols_2020 + numeric_pure_cols_2020\
                + categorical_cols_2020 + binary_cols_2020 + half_cols_2020

all_cols_2020에 없는 열

In [53]:
no_match_cols_2020 = [col for col in know_train[3].columns if col not in all_cols_2020]
no_match_cols_2020

['idx', 'knowcode', 'text_response', 'major', 'description']

In [54]:
know_train[3][[col for col in know_train[3].columns if col not in all_cols_2020]]

idx  knowcode  ...   major       description
0     27127    304003  ...     간호학              보건교사
1     27128    221401  ...    기계공학               변리사
2     27129    815001  ...      기계  자동조립라인·산업용로봇 조작원
3     27130    304003  ...     보건학              보건교사
4     27131    140201  ...  건축설비공학          건축구조 기술자
...     ...       ...  ...     ...               ...
8117  35244    121105  ...    지구과학            지질학연구원
8118  35245     26102  ...     광고학     광고·홍보·마케팅 사무원
8119  35246     31301  ...      국문             보험계리사
8120  35247    154105  ...    화학공학    의약품공학기술자 및 연구원
8121  35248     14201  ...     조리학         음식서비스 관리자

[8122 rows x 5 columns]

all_nontext_cols_2020의 열들은 다 2020에 있는가?

In [55]:
know_train[3][all_nontext_cols_2020]

saq1_1  saq1_2  saq2_1  saq2_2  saq3_1  ...  bq28  bq6  bq12_2  bq12_3  bq12_4
0          4       4       4       4       4  ...     1    7       3       4       4
1          5       6       5       6       4  ...     1    4       5       5       4
2          3       4       3       4       3  ...     1    5       4       4       4
3          4       5       2       3       3  ...     1    4       3       3       3
4          5       6       4       5       3  ...     1    7       4       5       5
...      ...     ...     ...     ...     ...  ...   ...  ...     ...     ...     ...
8117       4       6       3       5       3  ...     1    3       3       4       4
8118       3       4       4       5       2  ...     1    2       2       4       3
8119       5       6       3       4       3  ...     1    7       3       3       4
8120       5       6       5       5       4  ...     1    6       3       3       4
8121       2       5       4       6       3  ...     1    4       3       4       4

[8122 rows x 176 columns]

### half_cols 전처리

In [56]:
for col in half_cols_2020:
    print(col)
    print(know_train[0][col].value_counts())

bq6
3    3439
6    1919
5    1448
4    1106
2    1053
1     521
Name: bq6, dtype: int64
bq12_2
3    4194
4    2088
0    1630
2    1113
5     284
1     177
Name: bq12_2, dtype: int64
bq12_3
3    3641
4    3371
0    1343
5     556
2     515
1      60
Name: bq12_3, dtype: int64
bq12_4
4    4603
3    2985
0     879
5     834
2     170
1      15
Name: bq12_4, dtype: int64


In [57]:
know_train[3]['bq6_mark']= 0
know_train[3]['bq6_mark'].loc[know_train[3]['bq6'] == 7] = 1
know_train[3]['bq12_2_mark']= 0
know_train[3]['bq12_2_mark'].loc[know_train[3]['bq12_2'] == 9] = 1
know_train[3]['bq12_3_mark']= 0
know_train[3]['bq12_3_mark'].loc[know_train[3]['bq12_3'] == 9] = 1
know_train[3]['bq12_4_mark']= 0
know_train[3]['bq12_4_mark'].loc[know_train[3]['bq12_4'] == 9] = 1

In [58]:
know_train[3].loc[know_train[3]['bq6'] == 7, 'bq6'] = 3 # median filling
know_train[3].loc[know_train[3]['bq12_2'] == 9, 'bq12_2'] = 3 # median filling
know_train[3].loc[know_train[3]['bq12_3'] == 9, 'bq12_3'] = 3 # median filling
know_train[3].loc[know_train[3]['bq12_4'] == 9, 'bq12_4'] = 3 # median filling

In [59]:
half_cols_mark_2020 = ['bq6_mark', 'bq12_2_mark', 'bq12_3_mark', 'bq12_4_mark',]

## Train Test Split 후 ET & Catboost HPO 평가

In [ ]:
from sklearn.ensemble import ExtraTreesClassifier

In [60]:
use_cols_2017 = [col for col in know_train[0] if col not in no_match_cols_2017]
use_cols_2018 = [col for col in know_train[1] if col not in no_match_cols_2018]
use_cols_2019 = [col for col in know_train[2] if col not in no_match_cols_2019]
use_cols_2020 = [col for col in know_train[3] if col not in no_match_cols_2020]

In [ ]:
use_cols_2017

['aq1_1',
 'aq1_2',
 'aq2_1',
 'aq2_2',
 'aq3_1',
 'aq3_2',
 'aq4_1',
 'aq4_2',
 'aq5_1',
 'aq5_2',
 'aq6_1',
 'aq6_2',
 'aq7_1',
 'aq7_2',
 'aq8_1',
 'aq8_2',
 'aq9_1',
 'aq9_2',
 'aq10_1',
 'aq10_2',
 'aq11_1',
 'aq11_2',
 'aq12_1',
 'aq12_2',
 'aq13_1',
 'aq13_2',
 'aq14_1',
 'aq14_2',
 'aq15_1',
 'aq15_2',
 'aq16_1',
 'aq16_2',
 'aq17_1',
 'aq17_2',
 'aq18_1',
 'aq18_2',
 'aq19_1',
 'aq19_2',
 'aq20_1',
 'aq20_2',
 'aq21_1',
 'aq21_2',
 'aq22_1',
 'aq22_2',
 'aq23_1',
 'aq23_2',
 'aq24_1',
 'aq24_2',
 'aq25_1',
 'aq25_2',
 'aq26_1',
 'aq26_2',
 'aq27_1',
 'aq27_2',
 'aq28_1',
 'aq28_2',
 'aq29_1',
 'aq29_2',
 'aq30_1',
 'aq30_2',
 'aq31_1',
 'aq31_2',
 'aq32_1',
 'aq32_2',
 'aq33_1',
 'aq33_2',
 'aq34_1',
 'aq34_2',
 'aq35_1',
 'aq35_2',
 'aq36_1',
 'aq36_2',
 'aq37_1',
 'aq37_2',
 'aq38_1',
 'aq38_2',
 'aq39_1',
 'aq39_2',
 'aq40_1',
 'aq40_2',
 'aq41_1',
 'aq41_2',
 'bq1',
 'bq2',
 'bq3',
 'bq4',
 'bq5',
 'bq5_1',
 'bq6',
 'bq7',
 'bq8_1',
 'bq8_2',
 'bq8_3',
 'bq9',
 'bq10',
 'b

### 2017

In [61]:
from sklearn.model_selection import train_test_split

In [62]:
X_train_2017, X_valid_2017, y_train_2017, y_valid_2017 = train_test_split(know_train[0][use_cols_2017], know_train[0]['knowcode'],
                                                                        test_size=0.2, random_state=RANDOM_STATE)

In [ ]:
clf2017 = ExtraTreesClassifier(n_estimators=100, random_state = RANDOM_STATE)
clf2017.fit(X_train_2017, y_train_2017)

In [ ]:
pred_2017 = clf2017.predict(X_valid_2017)

In [63]:
from sklearn.metrics import f1_score
f1_score(y_valid_2017, pred_2017, average='macro')

NameError: ignored

![image.png](attachment:image.png)

In [ ]:
clf2017_200 = ExtraTreesClassifier(n_estimators=200, random_state = RANDOM_STATE)
clf2017_200.fit(X_train_2017, y_train_2017)

In [ ]:
pred_2017_200 = clf2017_200.predict(X_valid_2017)

In [ ]:
from sklearn.metrics import f1_score
f1_score(y_valid_2017, pred_2017_200, average='macro')

![image.png](attachment:image.png)

In [ ]:
clf2017_300 = ExtraTreesClassifier(n_estimators=300, random_state = RANDOM_STATE)
clf2017_300.fit(X_train_2017, y_train_2017)

In [ ]:
pred_2017_300 = clf2017_300.predict(X_valid_2017)

In [ ]:
from sklearn.metrics import f1_score
f1_score(y_valid_2017, pred_2017_300, average='macro')

In [65]:
!pip3 install catboost

     |████████████████████████████████| 76.1 MB 61 kB/s 


In [66]:
from catboost import CatBoostClassifier

In [68]:
from sklearn.metrics import f1_score

In [67]:
cat_multiclass_f1 = CatBoostClassifier(iterations=50, 
                                        loss_function='MultiClass',
                                       random_state=RANDOM_STATE,
                                       task_type='GPU',
                                       eval_metric='TotalF1',
                                       )
cat_multiclass_f1.fit(X_train_2017, y_train_2017)

Learning rate set to 0.5
0:	learn: 0.0323341	total: 2.83s	remaining: 2m 18s
1:	learn: 0.0527369	total: 5.59s	remaining: 2m 14s
2:	learn: 0.0715064	total: 8.29s	remaining: 2m 9s
3:	learn: 0.1009902	total: 11s	remaining: 2m 6s
4:	learn: 0.1295583	total: 13.8s	remaining: 2m 4s
5:	learn: 0.1696684	total: 16.4s	remaining: 2m
6:	learn: 0.2114476	total: 19.2s	remaining: 1m 58s
7:	learn: 0.2511576	total: 21.7s	remaining: 1m 54s
8:	learn: 0.2909695	total: 24.3s	remaining: 1m 50s
9:	learn: 0.3311083	total: 27s	remaining: 1m 48s
10:	learn: 0.3754867	total: 29.8s	remaining: 1m 45s
11:	learn: 0.4244916	total: 32.2s	remaining: 1m 41s
12:	learn: 0.4690994	total: 35s	remaining: 1m 39s
13:	learn: 0.5088831	total: 37.7s	remaining: 1m 36s
14:	learn: 0.5456360	total: 40.5s	remaining: 1m 34s
15:	learn: 0.5784585	total: 43.1s	remaining: 1m 31s
16:	learn: 0.6129070	total: 45.8s	remaining: 1m 28s
17:	learn: 0.6425935	total: 48.4s	remaining: 1m 26s
18:	learn: 0.6695600	total: 51.2s	remaining: 1m 23s
19:	learn:

In [70]:
# make the prediction using the resulting model
pred_2017_f1 = cat_multiclass_f1.predict(X_valid_2017)
f1_score(y_valid_2017, pred_2017_f1, average='macro')

0.3518584319782506

In [76]:
cat_2017_f1_400= CatBoostClassifier(iterations=400,
                                   #learning_rate=0.3,
                                   loss_function='MultiClass',
                                   random_state=RANDOM_STATE,
                                   task_type='GPU',
                                   #eval_metric='TotalF1',
                                  )
cat_2017_f1_400.fit(X_train_2017, y_train_2017)

Learning rate set to 0.21053
0:	learn: 6.0140250	total: 2.64s	remaining: 17m 33s
1:	learn: 5.7460721	total: 5.15s	remaining: 17m 4s
2:	learn: 5.5552831	total: 7.83s	remaining: 17m 16s
3:	learn: 5.4069311	total: 10.4s	remaining: 17m 10s
4:	learn: 5.2828129	total: 12.9s	remaining: 16m 58s
5:	learn: 5.1270592	total: 15.5s	remaining: 16m 58s
6:	learn: 5.0088678	total: 18s	remaining: 16m 52s
7:	learn: 4.8983768	total: 20.6s	remaining: 16m 49s
8:	learn: 4.7565729	total: 23.2s	remaining: 16m 46s
9:	learn: 4.6540688	total: 25.7s	remaining: 16m 43s
10:	learn: 4.5855391	total: 28.3s	remaining: 16m 41s
11:	learn: 4.4430876	total: 30.7s	remaining: 16m 32s
12:	learn: 4.3627711	total: 33.2s	remaining: 16m 28s
13:	learn: 4.2295856	total: 35.8s	remaining: 16m 27s
14:	learn: 4.1210039	total: 38.5s	remaining: 16m 27s
15:	learn: 4.0596558	total: 41s	remaining: 16m 25s
16:	learn: 3.9741216	total: 43.7s	remaining: 16m 23s
17:	learn: 3.8853026	total: 46.2s	remaining: 16m 19s
18:	learn: 3.8053040	total: 48.7

In [77]:
# make the prediction using the resulting model
pred_2017_f1_400 = cat_2017_f1_400.predict(X_valid_2017)
f1_score(y_valid_2017, pred_2017_f1_400, average='macro')

0.46625208459171263

![image.png](attachment:image.png)

In [ ]:
cat_multiclass = CatBoostClassifier(iterations=100, loss_function='MultiClass')
cat_multiclass.fit(X_train_2017, y_train_2017)

Learning rate set to 0.5
0:	learn: 6.0252181	total: 14.6s	remaining: 24m 3s
1:	learn: 5.6479230	total: 28.9s	remaining: 23m 35s
2:	learn: 5.3552233	total: 42.4s	remaining: 22m 51s
3:	learn: 4.9716866	total: 55.8s	remaining: 22m 18s
4:	learn: 4.7567589	total: 1m 9s	remaining: 22m 6s
5:	learn: 4.5435565	total: 1m 23s	remaining: 21m 47s
6:	learn: 4.3646767	total: 1m 36s	remaining: 21m 28s
7:	learn: 4.1921997	total: 1m 50s	remaining: 21m 9s
8:	learn: 4.0000073	total: 2m 3s	remaining: 20m 52s
9:	learn: 3.8356729	total: 2m 17s	remaining: 20m 35s
10:	learn: 3.6468028	total: 2m 30s	remaining: 20m 20s
11:	learn: 3.4405008	total: 2m 44s	remaining: 20m 4s
12:	learn: 3.2728279	total: 2m 58s	remaining: 19m 55s
13:	learn: 3.1617612	total: 3m 12s	remaining: 19m 40s
14:	learn: 3.0052373	total: 3m 25s	remaining: 19m 25s
15:	learn: 2.8507316	total: 3m 39s	remaining: 19m 10s
16:	learn: 2.7043590	total: 3m 51s	remaining: 18m 52s
17:	learn: 2.5841176	total: 4m 5s	remaining: 18m 39s
18:	learn: 2.4581683	tot

In [ ]:
# make the prediction using the resulting model
pred_2017_cat_multiclass = cat_multiclass.predict(X_valid_2017)
prob_2017_cat_multiclass = cat_multiclass.predict_proba(X_valid_2017)

In [ ]:
from sklearn.metrics import f1_score
f1_score(y_valid_2017, pred_2017_cat_multiclass, average='macro')

0.4037574227665002

![image.png](attachment:image.png)

In [ ]:
cat_multiclass_200 = CatBoostClassifier(iterations=200, loss_function='MultiClass')
cat_multiclass_200.fit(X_train_2017, y_train_2017)

Learning rate set to 0.341853
0:	learn: 6.0417251	total: 13.5s	remaining: 44m 48s
1:	learn: 5.8006068	total: 27.9s	remaining: 46m 5s
2:	learn: 5.5876688	total: 42.5s	remaining: 46m 30s
3:	learn: 5.3023286	total: 56.3s	remaining: 45m 57s
4:	learn: 5.1167141	total: 1m 10s	remaining: 45m 51s
5:	learn: 4.9489414	total: 1m 24s	remaining: 45m 24s
6:	learn: 4.7883004	total: 1m 37s	remaining: 44m 51s
7:	learn: 4.6516191	total: 1m 51s	remaining: 44m 29s
8:	learn: 4.4738417	total: 2m 5s	remaining: 44m 17s
9:	learn: 4.3439121	total: 2m 18s	remaining: 43m 56s
10:	learn: 4.1870338	total: 2m 32s	remaining: 43m 33s
11:	learn: 4.0218418	total: 2m 46s	remaining: 43m 24s
12:	learn: 3.8720959	total: 3m	remaining: 43m 17s
13:	learn: 3.7645984	total: 3m 14s	remaining: 42m 58s
14:	learn: 3.6358640	total: 3m 27s	remaining: 42m 39s
15:	learn: 3.5118171	total: 3m 40s	remaining: 42m 17s
16:	learn: 3.3741316	total: 3m 53s	remaining: 41m 52s
17:	learn: 3.2790161	total: 4m 6s	remaining: 41m 31s
18:	learn: 3.169074

150:	learn: 0.1765368	total: 32m 58s	remaining: 10m 42s
151:	learn: 0.1748141	total: 33m 11s	remaining: 10m 28s
152:	learn: 0.1725707	total: 33m 24s	remaining: 10m 15s
153:	learn: 0.1707041	total: 33m 37s	remaining: 10m 2s
154:	learn: 0.1683488	total: 33m 50s	remaining: 9m 49s
155:	learn: 0.1659156	total: 34m 5s	remaining: 9m 36s
156:	learn: 0.1633772	total: 34m 18s	remaining: 9m 23s
157:	learn: 0.1619649	total: 34m 30s	remaining: 9m 10s
158:	learn: 0.1603198	total: 34m 42s	remaining: 8m 57s
159:	learn: 0.1585494	total: 34m 56s	remaining: 8m 44s
160:	learn: 0.1563466	total: 35m 12s	remaining: 8m 31s
161:	learn: 0.1548301	total: 35m 25s	remaining: 8m 18s
162:	learn: 0.1528454	total: 35m 38s	remaining: 8m 5s
163:	learn: 0.1514549	total: 35m 52s	remaining: 7m 52s
164:	learn: 0.1498183	total: 36m 5s	remaining: 7m 39s
165:	learn: 0.1483681	total: 36m 18s	remaining: 7m 26s
166:	learn: 0.1462051	total: 36m 31s	remaining: 7m 13s
167:	learn: 0.1452375	total: 36m 43s	remaining: 6m 59s
168:	learn

In [ ]:
# make the prediction using the resulting model
pred_2017_cat_multiclass_200 = cat_multiclass_200.predict(X_valid_2017)
prob_2017_cat_multiclass_200 = cat_multiclass_200.predict_proba(X_valid_2017)

In [ ]:
from sklearn.metrics import f1_score
f1_score(y_valid_2017, pred_2017_cat_multiclass_200, average='macro')

0.4504248940227167

![image.png](attachment:image.png)

In [ ]:
cat_multiclass_200 = CatBoostClassifier(iterations=200, 
                                        loss_function='MultiClass',
                                       random_state=RANDOM_STATE,
                                       task_type='gpu',
                                       eval_metric='f1',
                                       )
cat_multiclass_200.fit(X_train_2017, y_train_2017)

### Trainset shape 정리

 + 2017: (9486, 155)
 + 2018: (9072, 130)
 + 2019: (8555, 152)
 + 2020: (8122, 185)

## Testset 불러오기

마찬가지로 년도별로 DataFrame으로 불러온 후 리스트에 할당합니다.


In [78]:
TEST_PATH_2017 = DATA_PATH + "test/KNOW_2017_test.csv"
TEST_PATH_2018 = DATA_PATH + "test/KNOW_2018_test.csv"
TEST_PATH_2019 = DATA_PATH + "test/KNOW_2019_test.csv"
TEST_PATH_2020 = DATA_PATH + "test/KNOW_2020_test.csv"

TEST_PATHs = [TEST_PATH_2017, TEST_PATH_2018, TEST_PATH_2019, TEST_PATH_2020]

In [79]:
know_test = [pd.read_csv(path) for path in TEST_PATHs]
know_test[0].head() # 2017년도 test 샘플

idx  aq1_1  aq1_2  aq2_1  ...  sim_job  bef_job    able_job  major
0    0      3      4      2  ...       없다       없다          없다    비서학
1    1      5      5      3  ...       없다       없다          없다    농화학
2    2      5      5      5  ...       없다       없다  광선, 홍보 담당자   신문방송
3    3      4      5      5  ...       없다       없다          없다     화학
4    4      5      6      4  ...     편집기자     사진작가         리포터   광고홍보

[5 rows x 149 columns]

In [80]:
know_test[1]['bq222']

0       2
1       0
2       2
3       2
4       2
       ..
9064    2
9065    0
9066    0
9067    0
9068    0
Name: bq222, Length: 9069, dtype: int64

## Train과 똑같은 전처리

2018년 자료에 이상하게 원핫인코딩된 문항 수정

전처리: categorical column이어야 하는 bq22, bq23, bq24가 제멋대로 bq231, bq232.. 로 표기되어 있음... 원핫인코딩도 아니고 레이블인코딩도 아닌 무언가...

In [81]:
know_test[1][['bq221',
              'bq222',
              'bq223',
              'bq231',
              'bq232',
              'bq233',
              'bq234',
              'bq235',
              'bq241',
              'bq242',
              'bq243',
              'bq244',
              'bq245',]]

bq221  bq222  bq223  bq231  bq232  ...  bq241  bq242  bq243  bq244  bq245
0         0      2    0.0      0      2  ...      0      0      0      0      5
1         1      0    0.0      1      0  ...      1      0      0      0      0
2         0      2    0.0      0      0  ...      0      0      0      4      0
3         0      2    0.0      0      0  ...      0      0      0      0      5
4         0      2    0.0      0      0  ...      0      0      0      4      0
...     ...    ...    ...    ...    ...  ...    ...    ...    ...    ...    ...
9064      0      2    0.0      0      0  ...      0      0      0      0      5
9065      1      0    0.0      1      0  ...      0      0      0      0      5
9066      1      0    0.0      1      0  ...      0      2      0      0      0
9067      1      0    0.0      0      0  ...      0      0      0      0      5
9068      1      0    0.0      1      0  ...      1      0      0      0      0

[9069 rows x 13 columns]

In [82]:
know_test[1]['bq22'] =  know_test[1]['bq221'] + know_test[1]['bq222'] + know_test[1]['bq223']

In [83]:
know_test[1]['bq23'] =  know_test[1]['bq231'] + know_test[1]['bq232'] + know_test[1]['bq233']+\
                         know_test[1]['bq234'] + know_test[1]['bq235']

In [84]:
know_test[1]['bq24'] =  know_test[1]['bq241'] + know_test[1]['bq242'] + know_test[1]['bq243']+\
                         know_test[1]['bq244'] + know_test[1]['bq245']

In [85]:
know_test[1].drop(['bq221',
              'bq222',
              'bq223',
              'bq231',
              'bq232',
              'bq233',
              'bq234',
              'bq235',
              'bq241',
              'bq242',
              'bq243',
              'bq244',
              'bq245',], axis=1, inplace=True)

half cols 전처리

In [86]:
for i in range(4):
    know_test[i]['bq6_mark']= 0
    know_test[i]['bq6_mark'].loc[know_test[i]['bq6'] == 7] = 1
    know_test[i]['bq12_2_mark']= 0
    know_test[i]['bq12_2_mark'].loc[know_test[i]['bq12_2'] == 9] = 1
    know_test[i]['bq12_3_mark']= 0
    know_test[i]['bq12_3_mark'].loc[know_test[i]['bq12_3'] == 9] = 1
    know_test[i]['bq12_4_mark']= 0
    know_test[i]['bq12_4_mark'].loc[know_test[i]['bq12_4'] == 9] = 1

In [87]:
for i in range(4):
    know_test[i].loc[know_test[i]['bq6'] == 7, 'bq6'] = 3 # median filling
    know_test[i].loc[know_test[i]['bq12_2'] == 9, 'bq12_2'] = 3 # median filling
    know_test[i].loc[know_test[i]['bq12_3'] == 9, 'bq12_3'] = 3 # median filling
    know_test[i].loc[know_test[i]['bq12_4'] == 9, 'bq12_4'] = 3 # median filling

In [88]:
for i in range(4):
    print("Train Size")
    print(know_train[i].shape)
    print("Test Size")
    print(know_test[i].shape)

Train Size
(9486, 155)
Test Size
(9486, 153)
Train Size
(9072, 130)
Test Size
(9069, 128)
Train Size
(8555, 152)
Test Size
(8554, 150)
Train Size
(8122, 185)
Test Size
(8122, 183)


2 차이는 'description'과 'knowcode'

## Catboost 학습

In [90]:
cat_main_2017 = CatBoostClassifier(iterations=400, 
                                   loss_function='MultiClass',
                                   random_state=RANDOM_STATE,
                                   task_type='GPU',
                                  )
cat_main_2017 .fit(know_train[0][use_cols_2017], know_train[0]['knowcode'])

Learning rate set to 0.22034
0:	learn: 6.0231999	total: 2.83s	remaining: 18m 49s
1:	learn: 5.7354794	total: 5.46s	remaining: 18m 6s
2:	learn: 5.5102573	total: 8.03s	remaining: 17m 43s
3:	learn: 5.3440255	total: 10.7s	remaining: 17m 38s
4:	learn: 5.1769417	total: 13.4s	remaining: 17m 36s
5:	learn: 5.0231422	total: 16.1s	remaining: 17m 34s
6:	learn: 4.8834067	total: 18.7s	remaining: 17m 30s
7:	learn: 4.7146667	total: 21.4s	remaining: 17m 26s
8:	learn: 4.6165900	total: 23.9s	remaining: 17m 16s
9:	learn: 4.4591351	total: 26.4s	remaining: 17m 7s
10:	learn: 4.3535432	total: 28.9s	remaining: 17m 1s
11:	learn: 4.2326076	total: 31.3s	remaining: 16m 51s
12:	learn: 4.1489662	total: 33.9s	remaining: 16m 47s
13:	learn: 4.0301571	total: 36.4s	remaining: 16m 44s
14:	learn: 3.8920832	total: 39s	remaining: 16m 41s
15:	learn: 3.8265334	total: 41.6s	remaining: 16m 38s
16:	learn: 3.7312878	total: 44.2s	remaining: 16m 36s
17:	learn: 3.6382605	total: 46.8s	remaining: 16m 32s
18:	learn: 3.5663124	total: 49.3

In [91]:
cat_main_2018 = CatBoostClassifier(iterations=400, 
                                   loss_function='MultiClass',
                                   random_state=RANDOM_STATE,
                                   task_type='GPU',
                                  )
cat_main_2018.fit(know_train[1][use_cols_2018], know_train[1]['knowcode'])

Learning rate set to 0.218343
0:	learn: 6.1712222	total: 2.88s	remaining: 19m 10s
1:	learn: 5.9276496	total: 5.71s	remaining: 18m 56s
2:	learn: 5.7886362	total: 8.62s	remaining: 19m
3:	learn: 5.6476078	total: 11.4s	remaining: 18m 52s
4:	learn: 5.5213707	total: 14.3s	remaining: 18m 49s
5:	learn: 5.4095138	total: 17.2s	remaining: 18m 48s
6:	learn: 5.2843679	total: 19.9s	remaining: 18m 39s
7:	learn: 5.1582582	total: 22.7s	remaining: 18m 34s
8:	learn: 5.0727995	total: 25.6s	remaining: 18m 31s
9:	learn: 5.0041353	total: 28.4s	remaining: 18m 27s
10:	learn: 4.9088628	total: 31.3s	remaining: 18m 27s
11:	learn: 4.7393866	total: 34.5s	remaining: 18m 36s
12:	learn: 4.6268498	total: 37.5s	remaining: 18m 36s
13:	learn: 4.5110449	total: 40.3s	remaining: 18m 32s
14:	learn: 4.3995527	total: 43.6s	remaining: 18m 37s
15:	learn: 4.2250865	total: 46.8s	remaining: 18m 42s
16:	learn: 4.0938977	total: 50.1s	remaining: 18m 49s
17:	learn: 3.9663689	total: 53.3s	remaining: 18m 51s
18:	learn: 3.8717534	total: 56

In [92]:
cat_main_2019 = CatBoostClassifier(iterations=400, 
                                   loss_function='MultiClass',
                                   random_state=RANDOM_STATE,
                                   task_type='GPU',
                                  )
cat_main_2019.fit(know_train[2][use_cols_2019], know_train[2]['knowcode'])

Learning rate set to 0.215745
0:	learn: 6.1052866	total: 2.69s	remaining: 17m 55s
1:	learn: 5.8597138	total: 5.48s	remaining: 18m 10s
2:	learn: 5.6537930	total: 8.46s	remaining: 18m 40s
3:	learn: 5.4587042	total: 11.5s	remaining: 18m 55s
4:	learn: 5.2966732	total: 14.2s	remaining: 18m 45s
5:	learn: 5.1205093	total: 17.3s	remaining: 18m 55s
6:	learn: 4.9712951	total: 20.1s	remaining: 18m 46s
7:	learn: 4.8118722	total: 23s	remaining: 18m 45s
8:	learn: 4.6581805	total: 25.9s	remaining: 18m 46s
9:	learn: 4.5550861	total: 28.5s	remaining: 18m 33s
10:	learn: 4.3922953	total: 31.3s	remaining: 18m 25s
11:	learn: 4.2969033	total: 34.1s	remaining: 18m 21s
12:	learn: 4.1822596	total: 36.9s	remaining: 18m 18s
13:	learn: 4.0734685	total: 39.7s	remaining: 18m 15s
14:	learn: 3.9574993	total: 42.5s	remaining: 18m 11s
15:	learn: 3.8551112	total: 45.5s	remaining: 18m 11s
16:	learn: 3.7507680	total: 48.5s	remaining: 18m 11s
17:	learn: 3.6410145	total: 51.5s	remaining: 18m 12s
18:	learn: 3.5491740	total: 

In [93]:
cat_main_2020 = CatBoostClassifier(iterations=400, 
                                   loss_function='MultiClass',
                                   random_state=RANDOM_STATE,
                                   task_type='GPU',
                                  )
cat_main_2020.fit(know_train[3][use_cols_2020], know_train[3]['knowcode'])

Learning rate set to 0.213471
0:	learn: 6.1024580	total: 2.88s	remaining: 19m 11s
1:	learn: 5.8744479	total: 5.37s	remaining: 17m 48s
2:	learn: 5.7256020	total: 7.97s	remaining: 17m 35s
3:	learn: 5.5935187	total: 10.7s	remaining: 17m 39s
4:	learn: 5.4670850	total: 13.3s	remaining: 17m 31s
5:	learn: 5.3045524	total: 16s	remaining: 17m 30s
6:	learn: 5.1539083	total: 18.5s	remaining: 17m 20s
7:	learn: 5.0489907	total: 21s	remaining: 17m 9s
8:	learn: 4.9084306	total: 23.6s	remaining: 17m 5s
9:	learn: 4.8099072	total: 26.1s	remaining: 16m 59s
10:	learn: 4.6827406	total: 28.8s	remaining: 16m 58s
11:	learn: 4.5448214	total: 31.5s	remaining: 16m 56s
12:	learn: 4.4375818	total: 34s	remaining: 16m 51s
13:	learn: 4.3499181	total: 36.6s	remaining: 16m 47s
14:	learn: 4.2473375	total: 39.1s	remaining: 16m 43s
15:	learn: 4.1174693	total: 41.7s	remaining: 16m 40s
16:	learn: 4.0060640	total: 44.3s	remaining: 16m 38s
17:	learn: 3.8979359	total: 47s	remaining: 16m 36s
18:	learn: 3.8089013	total: 49.5s	re

## 테스트셋 추출 및 학습

ID 값을 제외한 나머지 데이터를 이용하여 모델에 넣어 추론합니다.

In [94]:
use_cols_2017 = [col for col in know_train[0] if col not in no_match_cols_2017]
use_cols_2018 = [col for col in know_train[1] if col not in no_match_cols_2018]
use_cols_2019 = [col for col in know_train[2] if col not in no_match_cols_2019]
use_cols_2020 = [col for col in know_train[3] if col not in no_match_cols_2020]

In [95]:
know_train[0][use_cols_2017].shape

(9486, 147)

In [96]:
know_test[0][use_cols_2017].shape

(9486, 147)

In [97]:
pred2017 = cat_main_2017.predict(know_test[0][use_cols_2017])

In [98]:
pred2018 = cat_main_2018.predict(know_test[1][use_cols_2018])

In [99]:
pred2019 = cat_main_2019.predict(know_test[2][use_cols_2019])

In [100]:
pred2020 = cat_main_2020.predict(know_test[3][use_cols_2020])

In [101]:
et_predicts = list(pred2017) + list(pred2018) + list(pred2019) + list(pred2020)

# 제출

In [106]:
et_predicts = [et[0] for et in et_predicts]

In [107]:
submission = pd.read_csv('../data_0103/sample_submission.csv') # sample submission 불러오기

In [108]:
submission['knowcode'] = et_predicts

submission.to_csv('../submission_files/catboost_iter_400_manual_preprocess_data0113.csv', index=False)